{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "70181fce-b580-404f-824c-f532f54796da",
   "metadata": {},
   "source": [
    "# MIPath Vignette\n",
    "\n",
    "### Package installation\n",
    "\n",
    "Install MIPath with\n",
    "\n",
    "`pip install mipathway`\n",
    "\n",
    "# Example use case\n",
    "\n",
    "Here, we will give an example of how to run a simple pathway analysis using **MIPath**.\n",
    "\n",
    "## Acquiring and preprocessing the data\n",
    "\n",
    "### Gene expression data\n",
    "\n",
    "For this example, we will use a single cell dataset available at [EBI Single Cell Expression Atlas](https://www.ebi.ac.uk/gxa/sc/home), more particularly a dataset on **haematopoiesis**, taken from the following publication:\n",
    "\n",
    "Ranzoni AM, Tangherloni A, Berest I, Riva SG, Myers B et al. (2021) Integrative Single-Cell RNA-Seq and ATAC-Seq Analysis of Human Developmental Hematopoiesis\n",
    "\n",
    "The package **Scanpy** is very useful for this purpose, and can be installed using:\n",
    "\n",
    "`pip install scanpy`\n",
    "\n",
    "We first import both **MIPath** and **Scanpy**\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5c8b03ec-780f-41d4-a04f-7d5aa1b3354d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import mipath as mp\n",
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1743a95-7391-455f-ba38-3dadf692aca6",
   "metadata": {},
   "source": [
    "We then use Scanpy to download the dataset directly from the Single Cell Expression Atlas, by using its accession number *E-MTAB-9067*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2c6616e6-69ed-4611-97f6-ffdb25df5587",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/gus/envs/mitest2/lib/python3.10/site-packages/anndata/__init__.py:51: FutureWarning: `anndata.read` is deprecated, use `anndata.read_h5ad` instead. `ad.read` will be removed in mid 2024.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "dataset_id = 'E-MTAB-9067'\n",
    "\n",
    "anndata = sc.datasets.ebi_expression_atlas(dataset_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5232f3f0-6367-4389-be4d-88614ffd4504",
   "metadata": {},
   "source": [
    "We will now preprocess the data, by filtering out cells with less than 300 expressed genes, and genes that are expressed in less than 3 cells.\n",
    "\n",
    "We then normalize the library sizes of each cell."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "fa0f73ec-c4f1-412c-bfe2-7876e2f429db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filtering\n",
    "sc.pp.filter_cells(anndata, min_genes=200)\n",
    "sc.pp.filter_genes(anndata, min_cells=3)\n",
    "\n",
    "# Library size correction\n",
    "sc.pp.normalize_total(anndata, target_sum=1e4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "125c8aa8-d23c-407e-8b76-30509780bc14",
   "metadata": {},
   "source": [
    "Scanpy uses [anndata](https://anndata.readthedocs.io/en/latest/) data structures. We want to extract pandas DataFrames from it, which can be used as inputs to **MIPath**. \n",
    "More specifically, we need one DataFrame containing the gene expression information, and one with the metadata that we want to compare against."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1b85c30a-a59d-4deb-bc54-af9e930da226",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = anndata.obs\n",
    "\n",
    "exp_data = anndata.to_df()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9d08d84-bf25-4c70-8183-40c81e90a201",
   "metadata": {},
   "source": [
    "### Pathway annotation data\n",
    "\n",
    "We also need a source of pathway annotation. This is most commonly done by using a **gmt** annotation file, such as the ones available at [MSigDB](https://www.gsea-msigdb.org/gsea/msigdb). \n",
    "\n",
    "One such file is included in this repo. This file however uses **HUGO gene names**, and we would like to use **Ensembl** annotation, as it matches our data (more on that later), so we provide the function with at dictionary that translates gene IDs for HUGO to Ensembl, created in [BioMart](https://www.ensembl.org/info/data/biomart/index.html)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9ff0f0de-2ef7-4e24-90b4-9c88b92efab8",
   "metadata": {},
   "outputs": [],
   "source": [
    "kegg_genesets = mp.parse_gmt(gmt_path = './data/c2.cp.kegg.v7.5.1.symbols.gmt', gene_id_dict_path = './data/name2ensembl_human.txt')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "623d1f9c-904b-42bb-9c4e-43ce3b5dc48e",
   "metadata": {},
   "source": [
    "Conveniently, **MIPath** also includes a function to download pathway annotation directly from [Reactome](https://www.reactome.org), simply by specifying the organism, and gene annotation scheme."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b22e9989-feeb-445a-8e26-87bd261ab500",
   "metadata": {},
   "outputs": [],
   "source": [
    "reactome_genesets = mp.get_reactome(organism = 'HSA', gene_anot = 'Ensembl')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b355b407-a9f0-440d-8f03-40a56fcf6540",
   "metadata": {},
   "source": [
    "### Structure of the data\n",
    "\n",
    "Now that we have all the data we need, let's have a look at its structure, starting with the *expression data*:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "25a1e4ca-b660-4042-8df7-da5f444a0a02",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <th>ENSG00000000971</th>\n",
       "      <th>ENSG00000001036</th>\n",
       "      <th>ENSG00000001084</th>\n",
       "      <th>ENSG00000001167</th>\n",
       "      <th>ENSG00000001460</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000289685</th>\n",
       "      <th>ENSG00000289690</th>\n",
       "      <th>ENSG00000289692</th>\n",
       "      <th>ENSG00000289694</th>\n",
       "      <th>ENSG00000289695</th>\n",
       "      <th>ENSG00000289697</th>\n",
       "      <th>ENSG00000289700</th>\n",
       "      <th>ENSG00000289701</th>\n",
       "      <th>ENSG00000289716</th>\n",
       "      <th>ENSG00000289718</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERR4147602</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.131484</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.131484</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4147603</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.226321</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4147604</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.932760</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.101724</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.203448</td>\n",
       "      <td>...</td>\n",
       "      <td>4.830823</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.855688</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4147605</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.546372</td>\n",
       "      <td>0.09934</td>\n",
       "      <td>0.04967</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.149011</td>\n",
       "      <td>0.397362</td>\n",
       "      <td>0.049670</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.198681</td>\n",
       "      <td>...</td>\n",
       "      <td>0.320914</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4147606</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.001502</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.609936</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32777 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            ENSG00000000003  ENSG00000000419  ENSG00000000457  \\\n",
       "ERR4147602              0.0         0.131484          0.00000   \n",
       "ERR4147603              0.0         0.000000          0.00000   \n",
       "ERR4147604              0.0         1.932760          0.00000   \n",
       "ERR4147605              0.0         0.546372          0.09934   \n",
       "ERR4147606              0.0         1.001502          0.00000   \n",
       "\n",
       "            ENSG00000000460  ENSG00000000938  ENSG00000000971  \\\n",
       "ERR4147602          0.00000              0.0         0.000000   \n",
       "ERR4147603          0.00000              0.0         0.000000   \n",
       "ERR4147604          0.00000              0.0         0.000000   \n",
       "ERR4147605          0.04967              0.0         0.149011   \n",
       "ERR4147606          0.00000              0.0         6.609936   \n",
       "\n",
       "            ENSG00000001036  ENSG00000001084  ENSG00000001167  \\\n",
       "ERR4147602         0.000000         0.000000              0.0   \n",
       "ERR4147603         0.000000         0.000000              0.0   \n",
       "ERR4147604         0.000000         0.101724              0.0   \n",
       "ERR4147605         0.397362         0.049670              0.0   \n",
       "ERR4147606         0.000000         0.000000              0.0   \n",
       "\n",
       "            ENSG00000001460  ...  ENSG00000289685  ENSG00000289690  \\\n",
       "ERR4147602         0.000000  ...         0.131484              0.0   \n",
       "ERR4147603         0.000000  ...         0.226321              0.0   \n",
       "ERR4147604         0.203448  ...         4.830823              0.0   \n",
       "ERR4147605         0.198681  ...         0.320914              0.0   \n",
       "ERR4147606         0.000000  ...         0.000000              0.0   \n",
       "\n",
       "            ENSG00000289692  ENSG00000289694  ENSG00000289695  \\\n",
       "ERR4147602              0.0         0.000000              0.0   \n",
       "ERR4147603              0.0         0.000000              0.0   \n",
       "ERR4147604              0.0         0.855688              0.0   \n",
       "ERR4147605              0.0         0.000000              0.0   \n",
       "ERR4147606              0.0         0.000000              0.0   \n",
       "\n",
       "            ENSG00000289697  ENSG00000289700  ENSG00000289701  \\\n",
       "ERR4147602              0.0              0.0              0.0   \n",
       "ERR4147603              0.0              0.0              0.0   \n",
       "ERR4147604              0.0              0.0              0.0   \n",
       "ERR4147605              0.0              0.0              0.0   \n",
       "ERR4147606              0.0              0.0              0.0   \n",
       "\n",
       "            ENSG00000289716  ENSG00000289718  \n",
       "ERR4147602              0.0              0.0  \n",
       "ERR4147603              0.0              0.0  \n",
       "ERR4147604              0.0              0.0  \n",
       "ERR4147605              0.0              0.0  \n",
       "ERR4147606              0.0              0.0  \n",
       "\n",
       "[5 rows x 32777 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exp_data.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0372799b-597d-4f80-a357-52d79cf6acee",
   "metadata": {},
   "source": [
    "We see that we have as row names individual IDs for each sample, and as column names gene IDs.\n",
    "\n",
    "Let's now see the *metadata*:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "04f6746f-2ec4-401e-803e-072c0352fdd9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Characteristic[organism]</th>\n",
       "      <th>Sample Characteristic Ontology Term[organism]</th>\n",
       "      <th>Sample Characteristic[individual]</th>\n",
       "      <th>Sample Characteristic Ontology Term[individual]</th>\n",
       "      <th>Sample Characteristic[developmental stage]</th>\n",
       "      <th>Sample Characteristic Ontology Term[developmental stage]</th>\n",
       "      <th>Sample Characteristic[age]</th>\n",
       "      <th>Sample Characteristic Ontology Term[age]</th>\n",
       "      <th>Sample Characteristic[sex]</th>\n",
       "      <th>Sample Characteristic Ontology Term[sex]</th>\n",
       "      <th>...</th>\n",
       "      <th>Factor Value Ontology Term[single cell identifier]</th>\n",
       "      <th>Factor Value[organism part]</th>\n",
       "      <th>Factor Value Ontology Term[organism part]</th>\n",
       "      <th>Factor Value[sampling site]</th>\n",
       "      <th>Factor Value Ontology Term[sampling site]</th>\n",
       "      <th>Factor Value[inferred cell type - authors labels]</th>\n",
       "      <th>Factor Value Ontology Term[inferred cell type - authors labels]</th>\n",
       "      <th>Factor Value[inferred cell type - ontology labels]</th>\n",
       "      <th>Factor Value Ontology Term[inferred cell type - ontology labels]</th>\n",
       "      <th>n_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERR4147602</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>Sample2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>late embryonic stage</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0007220</td>\n",
       "      <td>17 week</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000383</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bone marrow</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002371</td>\n",
       "      <td>hip</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0001464</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4147603</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>Sample2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>late embryonic stage</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0007220</td>\n",
       "      <td>17 week</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000383</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bone marrow</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002371</td>\n",
       "      <td>hip</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0001464</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4147604</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>Sample2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>late embryonic stage</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0007220</td>\n",
       "      <td>17 week</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000383</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bone marrow</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002371</td>\n",
       "      <td>hip</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0001464</td>\n",
       "      <td>erythroid progenitor cell</td>\n",
       "      <td>http://purl.obolibrary.org/obo/CL_0000038</td>\n",
       "      <td>erythroid progenitor cell</td>\n",
       "      <td>http://purl.obolibrary.org/obo/CL_0000038</td>\n",
       "      <td>3929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4147605</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>Sample2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>late embryonic stage</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0007220</td>\n",
       "      <td>17 week</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000383</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bone marrow</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002371</td>\n",
       "      <td>hip</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0001464</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4147606</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>Sample2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>late embryonic stage</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0007220</td>\n",
       "      <td>17 week</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000383</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bone marrow</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002371</td>\n",
       "      <td>hip</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0001464</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2137</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Sample Characteristic[organism]  \\\n",
       "ERR4147602                    Homo sapiens   \n",
       "ERR4147603                    Homo sapiens   \n",
       "ERR4147604                    Homo sapiens   \n",
       "ERR4147605                    Homo sapiens   \n",
       "ERR4147606                    Homo sapiens   \n",
       "\n",
       "            Sample Characteristic Ontology Term[organism]  \\\n",
       "ERR4147602  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "ERR4147603  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "ERR4147604  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "ERR4147605  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "ERR4147606  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "\n",
       "           Sample Characteristic[individual]  \\\n",
       "ERR4147602                           Sample2   \n",
       "ERR4147603                           Sample2   \n",
       "ERR4147604                           Sample2   \n",
       "ERR4147605                           Sample2   \n",
       "ERR4147606                           Sample2   \n",
       "\n",
       "            Sample Characteristic Ontology Term[individual]  \\\n",
       "ERR4147602                                              NaN   \n",
       "ERR4147603                                              NaN   \n",
       "ERR4147604                                              NaN   \n",
       "ERR4147605                                              NaN   \n",
       "ERR4147606                                              NaN   \n",
       "\n",
       "           Sample Characteristic[developmental stage]  \\\n",
       "ERR4147602                       late embryonic stage   \n",
       "ERR4147603                       late embryonic stage   \n",
       "ERR4147604                       late embryonic stage   \n",
       "ERR4147605                       late embryonic stage   \n",
       "ERR4147606                       late embryonic stage   \n",
       "\n",
       "           Sample Characteristic Ontology Term[developmental stage]  \\\n",
       "ERR4147602      http://purl.obolibrary.org/obo/UBERON_0007220         \n",
       "ERR4147603      http://purl.obolibrary.org/obo/UBERON_0007220         \n",
       "ERR4147604      http://purl.obolibrary.org/obo/UBERON_0007220         \n",
       "ERR4147605      http://purl.obolibrary.org/obo/UBERON_0007220         \n",
       "ERR4147606      http://purl.obolibrary.org/obo/UBERON_0007220         \n",
       "\n",
       "           Sample Characteristic[age]  \\\n",
       "ERR4147602                    17 week   \n",
       "ERR4147603                    17 week   \n",
       "ERR4147604                    17 week   \n",
       "ERR4147605                    17 week   \n",
       "ERR4147606                    17 week   \n",
       "\n",
       "            Sample Characteristic Ontology Term[age]  \\\n",
       "ERR4147602                                       NaN   \n",
       "ERR4147603                                       NaN   \n",
       "ERR4147604                                       NaN   \n",
       "ERR4147605                                       NaN   \n",
       "ERR4147606                                       NaN   \n",
       "\n",
       "           Sample Characteristic[sex]  \\\n",
       "ERR4147602                     female   \n",
       "ERR4147603                     female   \n",
       "ERR4147604                     female   \n",
       "ERR4147605                     female   \n",
       "ERR4147606                     female   \n",
       "\n",
       "               Sample Characteristic Ontology Term[sex]  ...  \\\n",
       "ERR4147602  http://purl.obolibrary.org/obo/PATO_0000383  ...   \n",
       "ERR4147603  http://purl.obolibrary.org/obo/PATO_0000383  ...   \n",
       "ERR4147604  http://purl.obolibrary.org/obo/PATO_0000383  ...   \n",
       "ERR4147605  http://purl.obolibrary.org/obo/PATO_0000383  ...   \n",
       "ERR4147606  http://purl.obolibrary.org/obo/PATO_0000383  ...   \n",
       "\n",
       "           Factor Value Ontology Term[single cell identifier]  \\\n",
       "ERR4147602                                                NaN   \n",
       "ERR4147603                                                NaN   \n",
       "ERR4147604                                                NaN   \n",
       "ERR4147605                                                NaN   \n",
       "ERR4147606                                                NaN   \n",
       "\n",
       "           Factor Value[organism part]  \\\n",
       "ERR4147602                 bone marrow   \n",
       "ERR4147603                 bone marrow   \n",
       "ERR4147604                 bone marrow   \n",
       "ERR4147605                 bone marrow   \n",
       "ERR4147606                 bone marrow   \n",
       "\n",
       "                Factor Value Ontology Term[organism part]  \\\n",
       "ERR4147602  http://purl.obolibrary.org/obo/UBERON_0002371   \n",
       "ERR4147603  http://purl.obolibrary.org/obo/UBERON_0002371   \n",
       "ERR4147604  http://purl.obolibrary.org/obo/UBERON_0002371   \n",
       "ERR4147605  http://purl.obolibrary.org/obo/UBERON_0002371   \n",
       "ERR4147606  http://purl.obolibrary.org/obo/UBERON_0002371   \n",
       "\n",
       "           Factor Value[sampling site]  \\\n",
       "ERR4147602                         hip   \n",
       "ERR4147603                         hip   \n",
       "ERR4147604                         hip   \n",
       "ERR4147605                         hip   \n",
       "ERR4147606                         hip   \n",
       "\n",
       "                Factor Value Ontology Term[sampling site]  \\\n",
       "ERR4147602  http://purl.obolibrary.org/obo/UBERON_0001464   \n",
       "ERR4147603  http://purl.obolibrary.org/obo/UBERON_0001464   \n",
       "ERR4147604  http://purl.obolibrary.org/obo/UBERON_0001464   \n",
       "ERR4147605  http://purl.obolibrary.org/obo/UBERON_0001464   \n",
       "ERR4147606  http://purl.obolibrary.org/obo/UBERON_0001464   \n",
       "\n",
       "           Factor Value[inferred cell type - authors labels]  \\\n",
       "ERR4147602                                               NaN   \n",
       "ERR4147603                                               NaN   \n",
       "ERR4147604                         erythroid progenitor cell   \n",
       "ERR4147605                                               NaN   \n",
       "ERR4147606                                               NaN   \n",
       "\n",
       "           Factor Value Ontology Term[inferred cell type - authors labels]  \\\n",
       "ERR4147602                                                NaN                \n",
       "ERR4147603                                                NaN                \n",
       "ERR4147604          http://purl.obolibrary.org/obo/CL_0000038                \n",
       "ERR4147605                                                NaN                \n",
       "ERR4147606                                                NaN                \n",
       "\n",
       "            Factor Value[inferred cell type - ontology labels]  \\\n",
       "ERR4147602                                                NaN    \n",
       "ERR4147603                                                NaN    \n",
       "ERR4147604                          erythroid progenitor cell    \n",
       "ERR4147605                                                NaN    \n",
       "ERR4147606                                                NaN    \n",
       "\n",
       "           Factor Value Ontology Term[inferred cell type - ontology labels]  \\\n",
       "ERR4147602                                                NaN                 \n",
       "ERR4147603                                                NaN                 \n",
       "ERR4147604          http://purl.obolibrary.org/obo/CL_0000038                 \n",
       "ERR4147605                                                NaN                 \n",
       "ERR4147606                                                NaN                 \n",
       "\n",
       "            n_genes  \n",
       "ERR4147602     2191  \n",
       "ERR4147603     1810  \n",
       "ERR4147604     3929  \n",
       "ERR4147605     5115  \n",
       "ERR4147606     2137  \n",
       "\n",
       "[5 rows x 33 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcbc04e8-e84a-4553-9eb3-2944557ecff0",
   "metadata": {},
   "source": [
    "We see that we have a lot of information on our samples. Most importantly, **the row names match the ones in the expression data**.\n",
    "\n",
    "We also have many columns, and we do not want to use all of them. This is OK, we will specify the which columns to use later. Importantly, the columns we want to use contain *factors* that separate the samples, and not continuous data. If we want to analyse continuous metadata, it has to be discretized first in some way.\n",
    "\n",
    "We can also look at how the geneset DataFrame looks like:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a5b01478-6c7e-4444-9171-846c1c067363",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>url</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KEGG_N_GLYCAN_BIOSYNTHESIS</th>\n",
       "      <td>[ENSG00000284501, ENSG00000168282, ENSG0000003...</td>\n",
       "      <td>http://www.gsea-msigdb.org/gsea/msigdb/cards/K...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_OTHER_GLYCAN_DEGRADATION</th>\n",
       "      <td>[ENSG00000262446, ENSG00000277926, ENSG0000022...</td>\n",
       "      <td>http://www.gsea-msigdb.org/gsea/msigdb/cards/K...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_O_GLYCAN_BIOSYNTHESIS</th>\n",
       "      <td>[ENSG00000288490, ENSG00000158470, ENSG0000011...</td>\n",
       "      <td>http://www.gsea-msigdb.org/gsea/msigdb/cards/K...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_GLYCOSAMINOGLYCAN_DEGRADATION</th>\n",
       "      <td>[ENSG00000127415, ENSG00000010404, ENSG0000016...</td>\n",
       "      <td>http://www.gsea-msigdb.org/gsea/msigdb/cards/K...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_KERATAN_SULFATE</th>\n",
       "      <td>[ENSG00000033170, ENSG00000175264, ENSG0000008...</td>\n",
       "      <td>http://www.gsea-msigdb.org/gsea/msigdb/cards/K...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                genes  \\\n",
       "KEGG_N_GLYCAN_BIOSYNTHESIS                          [ENSG00000284501, ENSG00000168282, ENSG0000003...   \n",
       "KEGG_OTHER_GLYCAN_DEGRADATION                       [ENSG00000262446, ENSG00000277926, ENSG0000022...   \n",
       "KEGG_O_GLYCAN_BIOSYNTHESIS                          [ENSG00000288490, ENSG00000158470, ENSG0000011...   \n",
       "KEGG_GLYCOSAMINOGLYCAN_DEGRADATION                  [ENSG00000127415, ENSG00000010404, ENSG0000016...   \n",
       "KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_KERATAN_SUL...  [ENSG00000033170, ENSG00000175264, ENSG0000008...   \n",
       "\n",
       "                                                                                                  url  \n",
       "KEGG_N_GLYCAN_BIOSYNTHESIS                          http://www.gsea-msigdb.org/gsea/msigdb/cards/K...  \n",
       "KEGG_OTHER_GLYCAN_DEGRADATION                       http://www.gsea-msigdb.org/gsea/msigdb/cards/K...  \n",
       "KEGG_O_GLYCAN_BIOSYNTHESIS                          http://www.gsea-msigdb.org/gsea/msigdb/cards/K...  \n",
       "KEGG_GLYCOSAMINOGLYCAN_DEGRADATION                  http://www.gsea-msigdb.org/gsea/msigdb/cards/K...  \n",
       "KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_KERATAN_SUL...  http://www.gsea-msigdb.org/gsea/msigdb/cards/K...  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kegg_genesets.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fb5b9e8-39f3-42f4-89c2-86aeaa837705",
   "metadata": {},
   "source": [
    "It has pathway IDs as row names, and a column named *genes* that contains a list of genes in that pathway. This is all done by the parsing function.\n",
    "\n",
    "With all the data loaded, we can proceed with the MIPath analysis.\n",
    "\n",
    "## MIPath analysis\n",
    "\n",
    "The MIPath analysis is done in two separate steps:\n",
    "* **Decomposition**: decomposing the activity of a pathway in each cell by grouping cells with similar activities, and\n",
    "* **Scoring**: comparing the decomposed pathway activities with factors present in the metadata.\n",
    "\n",
    "The decomposition step is by far the most computationally intensive, and so it was kept separate, as its results can be saved and reused for different factors.\n",
    "\n",
    "### Decomposition step\n",
    "\n",
    "We use the `decompose_pathways` function, providing the expression data and pathway annotation:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fd014df2-ef9b-4cb4-9e96-01bb0802369b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Performing pathway decomposition\n",
      "(1/186) --- KEGG_N_GLYCAN_BIOSYNTHESIS --- 46 genes\n",
      "Computed 25 NN in 8.7 s\n",
      "SNN scores in 0.1 s\n",
      "Found 27 partitions in 1.0 s\n",
      "\n",
      "(2/186) --- KEGG_OTHER_GLYCAN_DEGRADATION --- 24 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 26 partitions in 0.5 s\n",
      "\n",
      "(3/186) --- KEGG_O_GLYCAN_BIOSYNTHESIS --- 30 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 30 partitions in 0.7 s\n",
      "\n",
      "(4/186) --- KEGG_GLYCOSAMINOGLYCAN_DEGRADATION --- 20 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 30 partitions in 0.7 s\n",
      "\n",
      "(5/186) --- KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_KERATAN_SULFATE --- 14 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 34 partitions in 1.0 s\n",
      "\n",
      "(6/186) --- KEGG_GLYCEROLIPID_METABOLISM --- 57 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 30 partitions in 1.0 s\n",
      "\n",
      "(7/186) --- KEGG_GLYCOSYLPHOSPHATIDYLINOSITOL_GPI_ANCHOR_BIOSYNTHESIS --- 25 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 26 partitions in 1.1 s\n",
      "\n",
      "(8/186) --- KEGG_GLYCEROPHOSPHOLIPID_METABOLISM --- 95 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 29 partitions in 0.7 s\n",
      "\n",
      "(9/186) --- KEGG_ETHER_LIPID_METABOLISM --- 34 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.2 s\n",
      "Found 27 partitions in 0.7 s\n",
      "\n",
      "(10/186) --- KEGG_ARACHIDONIC_ACID_METABOLISM --- 62 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 25 partitions in 1.3 s\n",
      "\n",
      "(11/186) --- KEGG_LINOLEIC_ACID_METABOLISM --- 29 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 32 partitions in 0.6 s\n",
      "\n",
      "(12/186) --- KEGG_ALPHA_LINOLENIC_ACID_METABOLISM --- 19 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 33 partitions in 0.5 s\n",
      "\n",
      "(13/186) --- KEGG_SPHINGOLIPID_METABOLISM --- 48 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 28 partitions in 0.8 s\n",
      "\n",
      "(14/186) --- KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_LACTO_AND_NEOLACTO_SERIES --- 27 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 28 partitions in 0.6 s\n",
      "\n",
      "(15/186) --- KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_GLOBO_SERIES --- 13 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 30 partitions in 0.8 s\n",
      "\n",
      "(16/186) --- KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_GANGLIO_SERIES --- 18 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 31 partitions in 1.1 s\n",
      "\n",
      "(17/186) --- KEGG_RIBOFLAVIN_METABOLISM --- 15 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 29 partitions in 0.6 s\n",
      "\n",
      "(18/186) --- KEGG_NICOTINATE_AND_NICOTINAMIDE_METABOLISM --- 23 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 24 partitions in 1.1 s\n",
      "\n",
      "(19/186) --- KEGG_PANTOTHENATE_AND_COA_BIOSYNTHESIS --- 16 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 32 partitions in 0.6 s\n",
      "\n",
      "(20/186) --- KEGG_AMINOACYL_TRNA_BIOSYNTHESIS --- 53 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 27 partitions in 1.2 s\n",
      "\n",
      "(21/186) --- KEGG_BASAL_TRANSCRIPTION_FACTORS --- 44 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 27 partitions in 1.0 s\n",
      "\n",
      "(22/186) --- KEGG_NON_HOMOLOGOUS_END_JOINING --- 12 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 26 partitions in 0.9 s\n",
      "\n",
      "(23/186) --- KEGG_SNARE_INTERACTIONS_IN_VESICULAR_TRANSPORT --- 37 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.2 s\n",
      "Found 31 partitions in 1.2 s\n",
      "\n",
      "(24/186) --- KEGG_LYSOSOME --- 137 genes\n",
      "Computed 25 NN in 0.4 s\n",
      "SNN scores in 0.2 s\n",
      "Found 17 partitions in 0.7 s\n",
      "\n",
      "(25/186) --- KEGG_CARDIAC_MUSCLE_CONTRACTION --- 81 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.1 s\n",
      "Found 17 partitions in 2.0 s\n",
      "\n",
      "(26/186) --- KEGG_RENIN_ANGIOTENSIN_SYSTEM --- 16 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 33 partitions in 0.9 s\n",
      "\n",
      "(27/186) --- KEGG_CIRCADIAN_RHYTHM_MAMMAL --- 12 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 30 partitions in 0.6 s\n",
      "\n",
      "(28/186) --- KEGG_TASTE_TRANSDUCTION --- 78 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.1 s\n",
      "Found 28 partitions in 0.6 s\n",
      "\n",
      "(29/186) --- KEGG_PROXIMAL_TUBULE_BICARBONATE_RECLAMATION --- 24 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.1 s\n",
      "Found 30 partitions in 1.1 s\n",
      "\n",
      "(30/186) --- KEGG_PATHOGENIC_ESCHERICHIA_COLI_INFECTION --- 67 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 14 partitions in 1.3 s\n",
      "\n",
      "(31/186) --- KEGG_SYSTEMIC_LUPUS_ERYTHEMATOSUS --- 237 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 19 partitions in 1.2 s\n",
      "\n",
      "(32/186) --- KEGG_PRIMARY_IMMUNODEFICIENCY --- 49 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 24 partitions in 0.9 s\n",
      "\n",
      "(33/186) --- KEGG_HYPERTROPHIC_CARDIOMYOPATHY_HCM --- 91 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 20 partitions in 0.8 s\n",
      "\n",
      "(34/186) --- KEGG_ARRHYTHMOGENIC_RIGHT_VENTRICULAR_CARDIOMYOPATHY_ARVC --- 76 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.1 s\n",
      "Found 20 partitions in 1.3 s\n",
      "\n",
      "(35/186) --- KEGG_DILATED_CARDIOMYOPATHY --- 100 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 18 partitions in 1.1 s\n",
      "\n",
      "(36/186) --- KEGG_GLYCOLYSIS_GLUCONEOGENESIS --- 65 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 18 partitions in 1.7 s\n",
      "\n",
      "(37/186) --- KEGG_CITRATE_CYCLE_TCA_CYCLE --- 31 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 22 partitions in 1.0 s\n",
      "\n",
      "(38/186) --- KEGG_PENTOSE_PHOSPHATE_PATHWAY --- 26 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 22 partitions in 1.4 s\n",
      "\n",
      "(39/186) --- KEGG_PENTOSE_AND_GLUCURONATE_INTERCONVERSIONS --- 29 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.2 s\n",
      "Found 30 partitions in 0.8 s\n",
      "\n",
      "(40/186) --- KEGG_FRUCTOSE_AND_MANNOSE_METABOLISM --- 34 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 25 partitions in 0.9 s\n",
      "\n",
      "(41/186) --- KEGG_GALACTOSE_METABOLISM --- 27 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 24 partitions in 1.4 s\n",
      "\n",
      "(42/186) --- KEGG_ASCORBATE_AND_ALDARATE_METABOLISM --- 26 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 37 partitions in 0.9 s\n",
      "\n",
      "(43/186) --- KEGG_FATTY_ACID_METABOLISM --- 41 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 26 partitions in 1.1 s\n",
      "\n",
      "(44/186) --- KEGG_STEROID_BIOSYNTHESIS --- 18 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 30 partitions in 0.8 s\n",
      "\n",
      "(45/186) --- KEGG_PRIMARY_BILE_ACID_BIOSYNTHESIS --- 16 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 30 partitions in 1.1 s\n",
      "\n",
      "(46/186) --- KEGG_STEROID_HORMONE_BIOSYNTHESIS --- 67 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 29 partitions in 1.0 s\n",
      "\n",
      "(47/186) --- KEGG_OXIDATIVE_PHOSPHORYLATION --- 155 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 14 partitions in 1.1 s\n",
      "\n",
      "(48/186) --- KEGG_PURINE_METABOLISM --- 164 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.2 s\n",
      "Found 19 partitions in 1.0 s\n",
      "\n",
      "(49/186) --- KEGG_PYRIMIDINE_METABOLISM --- 99 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.2 s\n",
      "Found 23 partitions in 1.4 s\n",
      "\n",
      "(50/186) --- KEGG_ALANINE_ASPARTATE_AND_GLUTAMATE_METABOLISM --- 32 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 29 partitions in 1.2 s\n",
      "\n",
      "(51/186) --- KEGG_GLYCINE_SERINE_AND_THREONINE_METABOLISM --- 31 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 31 partitions in 0.5 s\n",
      "\n",
      "(52/186) --- KEGG_CYSTEINE_AND_METHIONINE_METABOLISM --- 35 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 24 partitions in 1.4 s\n",
      "\n",
      "(53/186) --- KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_DEGRADATION --- 45 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 28 partitions in 1.2 s\n",
      "\n",
      "(54/186) --- KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_BIOSYNTHESIS --- 21 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.2 s\n",
      "Found 29 partitions in 0.7 s\n",
      "\n",
      "(55/186) --- KEGG_LYSINE_DEGRADATION --- 49 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 29 partitions in 1.1 s\n",
      "\n",
      "(56/186) --- KEGG_ARGININE_AND_PROLINE_METABOLISM --- 55 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 28 partitions in 1.2 s\n",
      "\n",
      "(57/186) --- KEGG_HISTIDINE_METABOLISM --- 29 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 27 partitions in 0.9 s\n",
      "\n",
      "(58/186) --- KEGG_TYROSINE_METABOLISM --- 43 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 25 partitions in 1.2 s\n",
      "\n",
      "(59/186) --- KEGG_PHENYLALANINE_METABOLISM --- 18 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 24 partitions in 1.1 s\n",
      "\n",
      "(60/186) --- KEGG_TRYPTOPHAN_METABOLISM --- 40 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 26 partitions in 0.9 s\n",
      "\n",
      "(61/186) --- KEGG_BETA_ALANINE_METABOLISM --- 21 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 26 partitions in 1.6 s\n",
      "\n",
      "(62/186) --- KEGG_TAURINE_AND_HYPOTAURINE_METABOLISM --- 10 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.1 s\n",
      "Found 54 partitions in 0.8 s\n",
      "\n",
      "(63/186) --- KEGG_SELENOAMINO_ACID_METABOLISM --- 25 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 27 partitions in 0.7 s\n",
      "\n",
      "(64/186) --- KEGG_GLUTATHIONE_METABOLISM --- 52 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 22 partitions in 1.4 s\n",
      "\n",
      "(65/186) --- KEGG_STARCH_AND_SUCROSE_METABOLISM --- 56 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 30 partitions in 0.9 s\n",
      "\n",
      "(66/186) --- KEGG_AMINO_SUGAR_AND_NUCLEOTIDE_SUGAR_METABOLISM --- 45 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 32 partitions in 0.8 s\n",
      "\n",
      "(67/186) --- KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_CHONDROITIN_SULFATE --- 22 genes\n",
      "Computed 25 NN in 0.4 s\n",
      "SNN scores in 0.1 s\n",
      "Found 27 partitions in 0.8 s\n",
      "\n",
      "(68/186) --- KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_HEPARAN_SULFATE --- 26 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 28 partitions in 0.5 s\n",
      "\n",
      "(69/186) --- KEGG_INOSITOL_PHOSPHATE_METABOLISM --- 56 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 25 partitions in 0.7 s\n",
      "\n",
      "(70/186) --- KEGG_PYRUVATE_METABOLISM --- 43 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 25 partitions in 1.1 s\n",
      "\n",
      "(71/186) --- KEGG_GLYOXYLATE_AND_DICARBOXYLATE_METABOLISM --- 15 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 28 partitions in 0.9 s\n",
      "\n",
      "(72/186) --- KEGG_PROPANOATE_METABOLISM --- 34 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 23 partitions in 1.4 s\n",
      "\n",
      "(73/186) --- KEGG_BUTANOATE_METABOLISM --- 34 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 26 partitions in 1.1 s\n",
      "\n",
      "(74/186) --- KEGG_ONE_CARBON_POOL_BY_FOLATE --- 19 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 25 partitions in 1.2 s\n",
      "\n",
      "(75/186) --- KEGG_FOLATE_BIOSYNTHESIS --- 10 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 45 partitions in 0.7 s\n",
      "\n",
      "(76/186) --- KEGG_RETINOL_METABOLISM --- 67 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 31 partitions in 1.1 s\n",
      "\n",
      "(77/186) --- KEGG_PORPHYRIN_AND_CHLOROPHYLL_METABOLISM --- 44 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 30 partitions in 1.0 s\n",
      "\n",
      "(78/186) --- KEGG_TERPENOID_BACKBONE_BIOSYNTHESIS --- 14 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 29 partitions in 0.7 s\n",
      "\n",
      "(79/186) --- KEGG_LIMONENE_AND_PINENE_DEGRADATION --- 9 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 32 partitions in 0.7 s\n",
      "\n",
      "(80/186) --- KEGG_NITROGEN_METABOLISM --- 23 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 29 partitions in 0.7 s\n",
      "\n",
      "(81/186) --- KEGG_SULFUR_METABOLISM --- 12 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 38 partitions in 0.8 s\n",
      "\n",
      "(82/186) --- KEGG_METABOLISM_OF_XENOBIOTICS_BY_CYTOCHROME_P450 --- 72 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 25 partitions in 1.6 s\n",
      "\n",
      "(83/186) --- KEGG_DRUG_METABOLISM_CYTOCHROME_P450 --- 79 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 26 partitions in 1.2 s\n",
      "\n",
      "(84/186) --- KEGG_DRUG_METABOLISM_OTHER_ENZYMES --- 54 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.2 s\n",
      "Found 26 partitions in 0.9 s\n",
      "\n",
      "(85/186) --- KEGG_BIOSYNTHESIS_OF_UNSATURATED_FATTY_ACIDS --- 22 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 26 partitions in 1.3 s\n",
      "\n",
      "(86/186) --- KEGG_ABC_TRANSPORTERS --- 60 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 31 partitions in 0.8 s\n",
      "\n",
      "(87/186) --- KEGG_RIBOSOME --- 104 genes\n",
      "Computed 25 NN in 0.4 s\n",
      "SNN scores in 0.1 s\n",
      "Found 17 partitions in 0.9 s\n",
      "\n",
      "(88/186) --- KEGG_RNA_DEGRADATION --- 79 genes\n",
      "Computed 25 NN in 0.4 s\n",
      "SNN scores in 0.1 s\n",
      "Found 20 partitions in 1.1 s\n",
      "\n",
      "(89/186) --- KEGG_RNA_POLYMERASE --- 30 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 27 partitions in 1.0 s\n",
      "\n",
      "(90/186) --- KEGG_DNA_REPLICATION --- 35 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 24 partitions in 1.6 s\n",
      "\n",
      "(91/186) --- KEGG_SPLICEOSOME --- 160 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.2 s\n",
      "Found 12 partitions in 1.0 s\n",
      "\n",
      "(92/186) --- KEGG_PROTEASOME --- 64 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.2 s\n",
      "Found 20 partitions in 0.9 s\n",
      "\n",
      "(93/186) --- KEGG_PROTEIN_EXPORT --- 23 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 24 partitions in 1.2 s\n",
      "\n",
      "(94/186) --- KEGG_PPAR_SIGNALING_PATHWAY --- 74 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 24 partitions in 1.2 s\n",
      "\n",
      "(95/186) --- KEGG_BASE_EXCISION_REPAIR --- 34 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 24 partitions in 1.0 s\n",
      "\n",
      "(96/186) --- KEGG_NUCLEOTIDE_EXCISION_REPAIR --- 52 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 29 partitions in 0.8 s\n",
      "\n",
      "(97/186) --- KEGG_MISMATCH_REPAIR --- 23 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 26 partitions in 1.0 s\n",
      "\n",
      "(98/186) --- KEGG_HOMOLOGOUS_RECOMBINATION --- 28 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.2 s\n",
      "Found 29 partitions in 0.7 s\n",
      "\n",
      "(99/186) --- KEGG_MAPK_SIGNALING_PATHWAY --- 307 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 17 partitions in 1.3 s\n",
      "\n",
      "(100/186) --- KEGG_ERBB_SIGNALING_PATHWAY --- 88 genes\n",
      "Computed 25 NN in 0.4 s\n",
      "SNN scores in 0.1 s\n",
      "Found 25 partitions in 1.3 s\n",
      "\n",
      "(101/186) --- KEGG_CALCIUM_SIGNALING_PATHWAY --- 181 genes\n",
      "Computed 25 NN in 0.4 s\n",
      "SNN scores in 0.1 s\n",
      "Found 14 partitions in 1.4 s\n",
      "\n",
      "(102/186) --- KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION --- 303 genes\n",
      "Computed 25 NN in 0.4 s\n",
      "SNN scores in 0.1 s\n",
      "Found 23 partitions in 0.8 s\n",
      "\n",
      "(103/186) --- KEGG_CHEMOKINE_SIGNALING_PATHWAY --- 213 genes\n",
      "Computed 25 NN in 0.4 s\n",
      "SNN scores in 0.2 s\n",
      "Found 22 partitions in 1.1 s\n",
      "\n",
      "(104/186) --- KEGG_PHOSPHATIDYLINOSITOL_SIGNALING_SYSTEM --- 80 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 23 partitions in 1.7 s\n",
      "\n",
      "(105/186) --- KEGG_NEUROACTIVE_LIGAND_RECEPTOR_INTERACTION --- 288 genes\n",
      "Computed 25 NN in 0.4 s\n",
      "SNN scores in 0.1 s\n",
      "Found 28 partitions in 0.6 s\n",
      "\n",
      "(106/186) --- KEGG_CELL_CYCLE --- 129 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.2 s\n",
      "Found 19 partitions in 0.9 s\n",
      "\n",
      "(107/186) --- KEGG_OOCYTE_MEIOSIS --- 118 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.2 s\n",
      "Found 15 partitions in 1.2 s\n",
      "\n",
      "(108/186) --- KEGG_P53_SIGNALING_PATHWAY --- 69 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 24 partitions in 1.0 s\n",
      "\n",
      "(109/186) --- KEGG_UBIQUITIN_MEDIATED_PROTEOLYSIS --- 137 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.2 s\n",
      "Found 16 partitions in 1.2 s\n",
      "\n",
      "(110/186) --- KEGG_REGULATION_OF_AUTOPHAGY --- 34 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 29 partitions in 1.2 s\n",
      "\n",
      "(111/186) --- KEGG_ENDOCYTOSIS --- 243 genes\n",
      "Computed 25 NN in 0.4 s\n",
      "SNN scores in 0.2 s\n",
      "Found 14 partitions in 0.7 s\n",
      "\n",
      "(112/186) --- KEGG_PEROXISOME --- 83 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 19 partitions in 1.1 s\n",
      "\n",
      "(113/186) --- KEGG_MTOR_SIGNALING_PATHWAY --- 53 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 16 partitions in 1.1 s\n",
      "\n",
      "(114/186) --- KEGG_APOPTOSIS --- 97 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.2 s\n",
      "Found 23 partitions in 0.8 s\n",
      "\n",
      "(115/186) --- KEGG_VASCULAR_SMOOTH_MUSCLE_CONTRACTION --- 119 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 17 partitions in 1.4 s\n",
      "\n",
      "(116/186) --- KEGG_WNT_SIGNALING_PATHWAY --- 164 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.2 s\n",
      "Found 14 partitions in 1.0 s\n",
      "\n",
      "(117/186) --- KEGG_DORSO_VENTRAL_AXIS_FORMATION --- 31 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 23 partitions in 0.9 s\n",
      "\n",
      "(118/186) --- KEGG_NOTCH_SIGNALING_PATHWAY --- 56 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 27 partitions in 1.4 s\n",
      "\n",
      "(119/186) --- KEGG_HEDGEHOG_SIGNALING_PATHWAY --- 59 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 27 partitions in 1.1 s\n",
      "\n",
      "(120/186) --- KEGG_TGF_BETA_SIGNALING_PATHWAY --- 94 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 20 partitions in 1.1 s\n",
      "\n",
      "(121/186) --- KEGG_AXON_GUIDANCE --- 133 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 22 partitions in 1.5 s\n",
      "\n",
      "(122/186) --- KEGG_VEGF_SIGNALING_PATHWAY --- 79 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 25 partitions in 0.8 s\n",
      "\n",
      "(123/186) --- KEGG_FOCAL_ADHESION --- 217 genes\n",
      "Computed 25 NN in 0.4 s\n",
      "SNN scores in 0.1 s\n",
      "Found 15 partitions in 1.2 s\n",
      "\n",
      "(124/186) --- KEGG_ECM_RECEPTOR_INTERACTION --- 105 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 25 partitions in 0.9 s\n",
      "\n",
      "(125/186) --- KEGG_CELL_ADHESION_MOLECULES_CAMS --- 250 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 18 partitions in 1.1 s\n",
      "\n",
      "(126/186) --- KEGG_ADHERENS_JUNCTION --- 80 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 15 partitions in 1.1 s\n",
      "\n",
      "(127/186) --- KEGG_TIGHT_JUNCTION --- 149 genes\n",
      "Computed 25 NN in 0.4 s\n",
      "SNN scores in 0.1 s\n",
      "Found 15 partitions in 0.8 s\n",
      "\n",
      "(128/186) --- KEGG_GAP_JUNCTION --- 103 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 19 partitions in 1.3 s\n",
      "\n",
      "(129/186) --- KEGG_COMPLEMENT_AND_COAGULATION_CASCADES --- 91 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.2 s\n",
      "Found 29 partitions in 0.6 s\n",
      "\n",
      "(130/186) --- KEGG_ANTIGEN_PROCESSING_AND_PRESENTATION --- 521 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 18 partitions in 1.4 s\n",
      "\n",
      "(131/186) --- KEGG_TOLL_LIKE_RECEPTOR_SIGNALING_PATHWAY --- 115 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 19 partitions in 1.7 s\n",
      "\n",
      "(132/186) --- KEGG_NOD_LIKE_RECEPTOR_SIGNALING_PATHWAY --- 72 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 20 partitions in 1.5 s\n",
      "\n",
      "(133/186) --- KEGG_RIG_I_LIKE_RECEPTOR_SIGNALING_PATHWAY --- 80 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 33 partitions in 1.1 s\n",
      "\n",
      "(134/186) --- KEGG_CYTOSOLIC_DNA_SENSING_PATHWAY --- 63 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 28 partitions in 1.0 s\n",
      "\n",
      "(135/186) --- KEGG_JAK_STAT_SIGNALING_PATHWAY --- 159 genes\n",
      "Computed 25 NN in 0.4 s\n",
      "SNN scores in 0.1 s\n",
      "Found 23 partitions in 0.8 s\n",
      "\n",
      "(136/186) --- KEGG_HEMATOPOIETIC_CELL_LINEAGE --- 103 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 19 partitions in 0.9 s\n",
      "\n",
      "(137/186) --- KEGG_NATURAL_KILLER_CELL_MEDIATED_CYTOTOXICITY --- 453 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.2 s\n",
      "Found 22 partitions in 1.0 s\n",
      "\n",
      "(138/186) --- KEGG_T_CELL_RECEPTOR_SIGNALING_PATHWAY --- 121 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 17 partitions in 0.8 s\n",
      "\n",
      "(139/186) --- KEGG_B_CELL_RECEPTOR_SIGNALING_PATHWAY --- 82 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 17 partitions in 0.7 s\n",
      "\n",
      "(140/186) --- KEGG_FC_EPSILON_RI_SIGNALING_PATHWAY --- 89 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 24 partitions in 1.1 s\n",
      "\n",
      "(141/186) --- KEGG_FC_GAMMA_R_MEDIATED_PHAGOCYTOSIS --- 100 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.2 s\n",
      "Found 19 partitions in 1.2 s\n",
      "\n",
      "(142/186) --- KEGG_LEUKOCYTE_TRANSENDOTHELIAL_MIGRATION --- 121 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.2 s\n",
      "Found 16 partitions in 1.7 s\n",
      "\n",
      "(143/186) --- KEGG_INTESTINAL_IMMUNE_NETWORK_FOR_IGA_PRODUCTION --- 114 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 20 partitions in 1.0 s\n",
      "\n",
      "(144/186) --- KEGG_LONG_TERM_POTENTIATION --- 73 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 23 partitions in 1.3 s\n",
      "\n",
      "(145/186) --- KEGG_NEUROTROPHIN_SIGNALING_PATHWAY --- 130 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.2 s\n",
      "Found 14 partitions in 0.9 s\n",
      "\n",
      "(146/186) --- KEGG_LONG_TERM_DEPRESSION --- 73 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 22 partitions in 1.6 s\n",
      "\n",
      "(147/186) --- KEGG_OLFACTORY_TRANSDUCTION --- 515 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 20 partitions in 0.9 s\n",
      "\n",
      "(148/186) --- KEGG_REGULATION_OF_ACTIN_CYTOSKELETON --- 219 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 15 partitions in 1.1 s\n",
      "\n",
      "(149/186) --- KEGG_INSULIN_SIGNALING_PATHWAY --- 154 genes\n",
      "Computed 25 NN in 0.4 s\n",
      "SNN scores in 0.2 s\n",
      "Found 16 partitions in 1.1 s\n",
      "\n",
      "(150/186) --- KEGG_GNRH_SIGNALING_PATHWAY --- 105 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 22 partitions in 1.1 s\n",
      "\n",
      "(151/186) --- KEGG_PROGESTERONE_MEDIATED_OOCYTE_MATURATION --- 89 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 17 partitions in 1.1 s\n",
      "\n",
      "(152/186) --- KEGG_MELANOGENESIS --- 106 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 20 partitions in 1.1 s\n",
      "\n",
      "(153/186) --- KEGG_ADIPOCYTOKINE_SIGNALING_PATHWAY --- 84 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 33 partitions in 1.5 s\n",
      "\n",
      "(154/186) --- KEGG_TYPE_II_DIABETES_MELLITUS --- 56 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 29 partitions in 1.2 s\n",
      "\n",
      "(155/186) --- KEGG_TYPE_I_DIABETES_MELLITUS --- 159 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 22 partitions in 1.0 s\n",
      "\n",
      "(156/186) --- KEGG_MATURITY_ONSET_DIABETES_OF_THE_YOUNG --- 25 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 35 partitions in 0.4 s\n",
      "\n",
      "(157/186) --- KEGG_ALDOSTERONE_REGULATED_SODIUM_REABSORPTION --- 41 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 28 partitions in 0.8 s\n",
      "\n",
      "(158/186) --- KEGG_VASOPRESSIN_REGULATED_WATER_REABSORPTION --- 46 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.2 s\n",
      "Found 19 partitions in 1.2 s\n",
      "\n",
      "(159/186) --- KEGG_ALZHEIMERS_DISEASE --- 191 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 13 partitions in 1.3 s\n",
      "\n",
      "(160/186) --- KEGG_PARKINSONS_DISEASE --- 146 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 13 partitions in 1.3 s\n",
      "\n",
      "(161/186) --- KEGG_AMYOTROPHIC_LATERAL_SCLEROSIS_ALS --- 64 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 23 partitions in 0.9 s\n",
      "\n",
      "(162/186) --- KEGG_HUNTINGTONS_DISEASE --- 202 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 14 partitions in 1.1 s\n",
      "\n",
      "(163/186) --- KEGG_PRION_DISEASES --- 40 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 24 partitions in 0.9 s\n",
      "\n",
      "(164/186) --- KEGG_VIBRIO_CHOLERAE_INFECTION --- 65 genes\n",
      "Computed 25 NN in 0.4 s\n",
      "SNN scores in 0.1 s\n",
      "Found 16 partitions in 0.9 s\n",
      "\n",
      "(165/186) --- KEGG_EPITHELIAL_CELL_SIGNALING_IN_HELICOBACTER_PYLORI_INFECTION --- 77 genes\n",
      "Computed 25 NN in 0.4 s\n",
      "SNN scores in 0.1 s\n",
      "Found 26 partitions in 1.5 s\n",
      "\n",
      "(166/186) --- KEGG_LEISHMANIA_INFECTION --- 147 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 22 partitions in 1.2 s\n",
      "\n",
      "(167/186) --- KEGG_PATHWAYS_IN_CANCER --- 337 genes\n",
      "Computed 25 NN in 0.4 s\n",
      "SNN scores in 0.2 s\n",
      "Found 16 partitions in 1.6 s\n",
      "\n",
      "(168/186) --- KEGG_COLORECTAL_CANCER --- 62 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 21 partitions in 1.3 s\n",
      "\n",
      "(169/186) --- KEGG_RENAL_CELL_CARCINOMA --- 71 genes\n",
      "Computed 25 NN in 0.4 s\n",
      "SNN scores in 0.1 s\n",
      "Found 30 partitions in 1.3 s\n",
      "\n",
      "(170/186) --- KEGG_PANCREATIC_CANCER --- 71 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 23 partitions in 1.2 s\n",
      "\n",
      "(171/186) --- KEGG_ENDOMETRIAL_CANCER --- 54 genes\n",
      "Computed 25 NN in 0.4 s\n",
      "SNN scores in 0.1 s\n",
      "Found 31 partitions in 0.9 s\n",
      "\n",
      "(172/186) --- KEGG_GLIOMA --- 68 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 19 partitions in 0.8 s\n",
      "\n",
      "(173/186) --- KEGG_PROSTATE_CANCER --- 92 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 18 partitions in 1.1 s\n",
      "\n",
      "(174/186) --- KEGG_THYROID_CANCER --- 34 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.2 s\n",
      "Found 27 partitions in 0.8 s\n",
      "\n",
      "(175/186) --- KEGG_BASAL_CELL_CARCINOMA --- 57 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 27 partitions in 0.8 s\n",
      "\n",
      "(176/186) --- KEGG_MELANOMA --- 75 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 29 partitions in 1.1 s\n",
      "\n",
      "(177/186) --- KEGG_BLADDER_CANCER --- 43 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 25 partitions in 1.3 s\n",
      "\n",
      "(178/186) --- KEGG_CHRONIC_MYELOID_LEUKEMIA --- 75 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 25 partitions in 0.9 s\n",
      "\n",
      "(179/186) --- KEGG_ACUTE_MYELOID_LEUKEMIA --- 58 genes\n",
      "Computed 25 NN in 0.4 s\n",
      "SNN scores in 0.1 s\n",
      "Found 28 partitions in 1.2 s\n",
      "\n",
      "(180/186) --- KEGG_SMALL_CELL_LUNG_CANCER --- 91 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 25 partitions in 1.2 s\n",
      "\n",
      "(181/186) --- KEGG_NON_SMALL_CELL_LUNG_CANCER --- 61 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 27 partitions in 1.0 s\n",
      "\n",
      "(182/186) --- KEGG_ASTHMA --- 102 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 25 partitions in 1.2 s\n",
      "\n",
      "(183/186) --- KEGG_AUTOIMMUNE_THYROID_DISEASE --- 155 genes\n",
      "Computed 25 NN in 0.4 s\n",
      "SNN scores in 0.1 s\n",
      "Found 22 partitions in 0.8 s\n",
      "\n",
      "(184/186) --- KEGG_ALLOGRAFT_REJECTION --- 146 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 21 partitions in 1.0 s\n",
      "\n",
      "(185/186) --- KEGG_GRAFT_VERSUS_HOST_DISEASE --- 302 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.1 s\n",
      "Found 19 partitions in 1.4 s\n",
      "\n",
      "(186/186) --- KEGG_VIRAL_MYOCARDITIS --- 173 genes\n",
      "Computed 25 NN in 0.3 s\n",
      "SNN scores in 0.2 s\n",
      "Found 17 partitions in 1.0 s\n",
      "\n"
     ]
    }
   ],
   "source": [
    "decomposed_df = mp.decompose_pathways(data = exp_data, gene_sets_df = kegg_genesets)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b17fa8f-7354-4c89-8104-58cbbfd57837",
   "metadata": {},
   "source": [
    "It gives us a new DataFrame, with the cells grouped by the activity of each pathway.\n",
    "\n",
    "### Scoring step\n",
    "\n",
    "We can now finally score the pathways. In this example, we want to analyse which pathways are most associated with different **cell types**, so we will use the `Factor Value[inferred cell type - authors labels]` column in our metadata. We are also curious to see if there is any pathway in this case that is associated with the different **sampling sites**, which can be found in the `Factor Value[sampling site]` column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b7c7a9d1-95bc-4d05-98d3-b6ac5ffc248c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Scoring Factor Value[inferred cell type - authors labels]\n",
      "Scoring Factor Value[sampling site]\n"
     ]
    }
   ],
   "source": [
    "factors_to_score = ['Factor Value[inferred cell type - authors labels]', 'Factor Value[sampling site]']\n",
    "\n",
    "result = mp.score_factors(decomposed_df = decomposed_df, metadata = metadata, factors = factors_to_score)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "526ed940-42a2-4eea-b769-95527e6fe169",
   "metadata": {},
   "source": [
    "### Results\n",
    "\n",
    "We can now display the results, sorting it by which pathways recieved the highest score for the **cell type**:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "21a5d13b-4f7f-42a1-99c0-1427c12528f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Factor Value[inferred cell type - authors labels]</th>\n",
       "      <th>Factor Value[sampling site]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KEGG_HEMATOPOIETIC_CELL_LINEAGE</th>\n",
       "      <td>0.326898</td>\n",
       "      <td>0.024368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_ANTIGEN_PROCESSING_AND_PRESENTATION</th>\n",
       "      <td>0.282982</td>\n",
       "      <td>0.042522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_LYSOSOME</th>\n",
       "      <td>0.262620</td>\n",
       "      <td>0.014868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION</th>\n",
       "      <td>0.256345</td>\n",
       "      <td>0.010223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_CELL_ADHESION_MOLECULES_CAMS</th>\n",
       "      <td>0.231899</td>\n",
       "      <td>0.048265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_JAK_STAT_SIGNALING_PATHWAY</th>\n",
       "      <td>0.213601</td>\n",
       "      <td>0.010626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_PRIMARY_IMMUNODEFICIENCY</th>\n",
       "      <td>0.205740</td>\n",
       "      <td>0.008597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_SYSTEMIC_LUPUS_ERYTHEMATOSUS</th>\n",
       "      <td>0.203573</td>\n",
       "      <td>0.031544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_B_CELL_RECEPTOR_SIGNALING_PATHWAY</th>\n",
       "      <td>0.197873</td>\n",
       "      <td>0.015963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_TYPE_I_DIABETES_MELLITUS</th>\n",
       "      <td>0.193569</td>\n",
       "      <td>0.041785</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Factor Value[inferred cell type - authors labels]  \\\n",
       "KEGG_HEMATOPOIETIC_CELL_LINEAGE                                                       0.326898   \n",
       "KEGG_ANTIGEN_PROCESSING_AND_PRESENTATION                                              0.282982   \n",
       "KEGG_LYSOSOME                                                                         0.262620   \n",
       "KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION                                           0.256345   \n",
       "KEGG_CELL_ADHESION_MOLECULES_CAMS                                                     0.231899   \n",
       "KEGG_JAK_STAT_SIGNALING_PATHWAY                                                       0.213601   \n",
       "KEGG_PRIMARY_IMMUNODEFICIENCY                                                         0.205740   \n",
       "KEGG_SYSTEMIC_LUPUS_ERYTHEMATOSUS                                                     0.203573   \n",
       "KEGG_B_CELL_RECEPTOR_SIGNALING_PATHWAY                                                0.197873   \n",
       "KEGG_TYPE_I_DIABETES_MELLITUS                                                         0.193569   \n",
       "\n",
       "                                             Factor Value[sampling site]  \n",
       "KEGG_HEMATOPOIETIC_CELL_LINEAGE                                 0.024368  \n",
       "KEGG_ANTIGEN_PROCESSING_AND_PRESENTATION                        0.042522  \n",
       "KEGG_LYSOSOME                                                   0.014868  \n",
       "KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION                     0.010223  \n",
       "KEGG_CELL_ADHESION_MOLECULES_CAMS                               0.048265  \n",
       "KEGG_JAK_STAT_SIGNALING_PATHWAY                                 0.010626  \n",
       "KEGG_PRIMARY_IMMUNODEFICIENCY                                   0.008597  \n",
       "KEGG_SYSTEMIC_LUPUS_ERYTHEMATOSUS                               0.031544  \n",
       "KEGG_B_CELL_RECEPTOR_SIGNALING_PATHWAY                          0.015963  \n",
       "KEGG_TYPE_I_DIABETES_MELLITUS                                   0.041785  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result.sort_values('Factor Value[inferred cell type - authors labels]', ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2eb5fbde-bf98-46de-b606-0502debf1826",
   "metadata": {},
   "source": [
    "We see that the pathway **KEGG_HEMATOPOIETIC_CELL_LINEAGE** has the highest score, which makes much sense."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
