{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7fd239cf",
   "metadata": {},
   "source": [
    "# MIPath Vignette\n",
    "\n",
    "## Package installation\n",
    "\n",
    "Install MIPath with\n",
    "\n",
    "`pip install mipathway`\n",
    "\n",
    "## Usage example\n",
    "\n",
    "Here we will give an example on how to run a standard pathway analysis with MIPath.\n",
    "\n",
    "First, we will load the libraries. Here we will also use **scanpy** to download and process a single cell dataset. You can install **scanpy** with `pip install scanpy`\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "80aae7f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import mipath\n",
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58d5efeb",
   "metadata": {},
   "source": [
    "We will use a dataset of a *hematopoietic cell lineage*, taken from the paper:\n",
    "\n",
    "Ranzoni AM, Tangherloni A, Berest I, Riva SG, Myers B et al. (2021) Integrative Single-Cell RNA-Seq and ATAC-Seq Analysis of Human Developmental Hematopoiesis\n",
    "\n",
    "It is present in the EBI Single Cell Expression Atlas, under the id E-MTAB-9067. We use **scanpy** to download the dataset, and preprocess it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "826a914d",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_id = 'E-MTAB-9067'\n",
    "\n",
    "adata = sc.datasets.ebi_expression_atlas(dataset_id)\n",
    "     \n",
    "# Basic filtering\n",
    "sc.pp.filter_cells(adata, min_genes=200)\n",
    "sc.pp.filter_genes(adata, min_cells=3)\n",
    "\n",
    "# Library size correction\n",
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "\n",
    "# Logarithmize the data\n",
    "sc.pp.log1p(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e4102d9",
   "metadata": {},
   "source": [
    "Next, we extract the two dataframes we need for the analysis, one with the gene expression information, and a second with phenotype information for each cell."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f97b55cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata = adata.obs\n",
    "data = adata.to_df()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8214fd7b",
   "metadata": {},
   "source": [
    "The dataframe with gene expression looks like this"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4ebd3d64",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <th>ENSG00000000971</th>\n",
       "      <th>ENSG00000001036</th>\n",
       "      <th>ENSG00000001084</th>\n",
       "      <th>ENSG00000001167</th>\n",
       "      <th>ENSG00000001460</th>\n",
       "      <th>...</th>\n",
       "      <th>ENSG00000289685</th>\n",
       "      <th>ENSG00000289690</th>\n",
       "      <th>ENSG00000289692</th>\n",
       "      <th>ENSG00000289694</th>\n",
       "      <th>ENSG00000289695</th>\n",
       "      <th>ENSG00000289697</th>\n",
       "      <th>ENSG00000289700</th>\n",
       "      <th>ENSG00000289701</th>\n",
       "      <th>ENSG00000289716</th>\n",
       "      <th>ENSG00000289718</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERR4147602</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.123530</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.123530</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4147603</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.204019</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4147604</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.075944</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.096876</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.185191</td>\n",
       "      <td>...</td>\n",
       "      <td>1.763158</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.618255</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4147605</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.435912</td>\n",
       "      <td>0.094710</td>\n",
       "      <td>0.048476</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.138901</td>\n",
       "      <td>0.334586</td>\n",
       "      <td>0.048476</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.181222</td>\n",
       "      <td>...</td>\n",
       "      <td>0.278324</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4147606</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693898</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.029455</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4147607</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.031404</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.027872</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.081388</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4147608</th>\n",
       "      <td>0.104638</td>\n",
       "      <td>0.036110</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.043839</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.036110</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.036110</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4147609</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.476586</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4147610</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.023072</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.398879</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4147611</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.101359</td>\n",
       "      <td>0.101359</td>\n",
       "      <td>0.913659</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.776297</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.642046</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 32777 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            ENSG00000000003  ENSG00000000419  ENSG00000000457  \\\n",
       "ERR4147602         0.000000         0.123530         0.000000   \n",
       "ERR4147603         0.000000         0.000000         0.000000   \n",
       "ERR4147604         0.000000         1.075944         0.000000   \n",
       "ERR4147605         0.000000         0.435912         0.094710   \n",
       "ERR4147606         0.000000         0.693898         0.000000   \n",
       "ERR4147607         0.000000         0.000000         0.000000   \n",
       "ERR4147608         0.104638         0.036110         0.000000   \n",
       "ERR4147609         0.000000         0.000000         0.000000   \n",
       "ERR4147610         0.000000         0.023072         0.000000   \n",
       "ERR4147611         0.000000         0.101359         0.101359   \n",
       "\n",
       "            ENSG00000000460  ENSG00000000938  ENSG00000000971  \\\n",
       "ERR4147602         0.000000              0.0         0.000000   \n",
       "ERR4147603         0.000000              0.0         0.000000   \n",
       "ERR4147604         0.000000              0.0         0.000000   \n",
       "ERR4147605         0.048476              0.0         0.138901   \n",
       "ERR4147606         0.000000              0.0         2.029455   \n",
       "ERR4147607         0.000000              0.0         0.031404   \n",
       "ERR4147608         0.043839              0.0         0.000000   \n",
       "ERR4147609         0.000000              0.0         0.476586   \n",
       "ERR4147610         0.000000              0.0         0.000000   \n",
       "ERR4147611         0.913659              0.0         0.000000   \n",
       "\n",
       "            ENSG00000001036  ENSG00000001084  ENSG00000001167  \\\n",
       "ERR4147602         0.000000         0.000000              0.0   \n",
       "ERR4147603         0.000000         0.000000              0.0   \n",
       "ERR4147604         0.000000         0.096876              0.0   \n",
       "ERR4147605         0.334586         0.048476              0.0   \n",
       "ERR4147606         0.000000         0.000000              0.0   \n",
       "ERR4147607         0.000000         0.027872              0.0   \n",
       "ERR4147608         0.000000         0.036110              0.0   \n",
       "ERR4147609         0.000000         0.000000              0.0   \n",
       "ERR4147610         0.000000         0.398879              0.0   \n",
       "ERR4147611         0.000000         0.000000              0.0   \n",
       "\n",
       "            ENSG00000001460  ...  ENSG00000289685  ENSG00000289690  \\\n",
       "ERR4147602         0.000000  ...         0.123530              0.0   \n",
       "ERR4147603         0.000000  ...         0.204019              0.0   \n",
       "ERR4147604         0.185191  ...         1.763158              0.0   \n",
       "ERR4147605         0.181222  ...         0.278324              0.0   \n",
       "ERR4147606         0.000000  ...         0.000000              0.0   \n",
       "ERR4147607         0.081388  ...         0.000000              0.0   \n",
       "ERR4147608         0.000000  ...         0.036110              0.0   \n",
       "ERR4147609         0.000000  ...         0.000000              0.0   \n",
       "ERR4147610         0.000000  ...         0.000000              0.0   \n",
       "ERR4147611         0.000000  ...         0.776297              0.0   \n",
       "\n",
       "            ENSG00000289692  ENSG00000289694  ENSG00000289695  \\\n",
       "ERR4147602              0.0         0.000000              0.0   \n",
       "ERR4147603              0.0         0.000000              0.0   \n",
       "ERR4147604              0.0         0.618255              0.0   \n",
       "ERR4147605              0.0         0.000000              0.0   \n",
       "ERR4147606              0.0         0.000000              0.0   \n",
       "ERR4147607              0.0         0.000000              0.0   \n",
       "ERR4147608              0.0         0.000000              0.0   \n",
       "ERR4147609              0.0         0.000000              0.0   \n",
       "ERR4147610              0.0         0.000000              0.0   \n",
       "ERR4147611              0.0         0.642046              0.0   \n",
       "\n",
       "            ENSG00000289697  ENSG00000289700  ENSG00000289701  \\\n",
       "ERR4147602              0.0              0.0              0.0   \n",
       "ERR4147603              0.0              0.0              0.0   \n",
       "ERR4147604              0.0              0.0              0.0   \n",
       "ERR4147605              0.0              0.0              0.0   \n",
       "ERR4147606              0.0              0.0              0.0   \n",
       "ERR4147607              0.0              0.0              0.0   \n",
       "ERR4147608              0.0              0.0              0.0   \n",
       "ERR4147609              0.0              0.0              0.0   \n",
       "ERR4147610              0.0              0.0              0.0   \n",
       "ERR4147611              0.0              0.0              0.0   \n",
       "\n",
       "            ENSG00000289716  ENSG00000289718  \n",
       "ERR4147602              0.0              0.0  \n",
       "ERR4147603              0.0              0.0  \n",
       "ERR4147604              0.0              0.0  \n",
       "ERR4147605              0.0              0.0  \n",
       "ERR4147606              0.0              0.0  \n",
       "ERR4147607              0.0              0.0  \n",
       "ERR4147608              0.0              0.0  \n",
       "ERR4147609              0.0              0.0  \n",
       "ERR4147610              0.0              0.0  \n",
       "ERR4147611              0.0              0.0  \n",
       "\n",
       "[10 rows x 32777 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24aa8ae3",
   "metadata": {},
   "source": [
    "And the one with phenotype information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a05b487b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample Characteristic[organism]</th>\n",
       "      <th>Sample Characteristic Ontology Term[organism]</th>\n",
       "      <th>Sample Characteristic[individual]</th>\n",
       "      <th>Sample Characteristic Ontology Term[individual]</th>\n",
       "      <th>Sample Characteristic[developmental stage]</th>\n",
       "      <th>Sample Characteristic Ontology Term[developmental stage]</th>\n",
       "      <th>Sample Characteristic[age]</th>\n",
       "      <th>Sample Characteristic Ontology Term[age]</th>\n",
       "      <th>Sample Characteristic[sex]</th>\n",
       "      <th>Sample Characteristic Ontology Term[sex]</th>\n",
       "      <th>...</th>\n",
       "      <th>Factor Value Ontology Term[single cell identifier]</th>\n",
       "      <th>Factor Value[organism part]</th>\n",
       "      <th>Factor Value Ontology Term[organism part]</th>\n",
       "      <th>Factor Value[sampling site]</th>\n",
       "      <th>Factor Value Ontology Term[sampling site]</th>\n",
       "      <th>Factor Value[inferred cell type - authors labels]</th>\n",
       "      <th>Factor Value Ontology Term[inferred cell type - authors labels]</th>\n",
       "      <th>Factor Value[inferred cell type - ontology labels]</th>\n",
       "      <th>Factor Value Ontology Term[inferred cell type - ontology labels]</th>\n",
       "      <th>n_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ERR4147602</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>Sample2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>late embryonic stage</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0007220</td>\n",
       "      <td>17 week</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000383</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bone marrow</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002371</td>\n",
       "      <td>hip</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0001464</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4147603</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>Sample2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>late embryonic stage</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0007220</td>\n",
       "      <td>17 week</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000383</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bone marrow</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002371</td>\n",
       "      <td>hip</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0001464</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4147604</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>Sample2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>late embryonic stage</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0007220</td>\n",
       "      <td>17 week</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000383</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bone marrow</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002371</td>\n",
       "      <td>hip</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0001464</td>\n",
       "      <td>erythroid progenitor cell</td>\n",
       "      <td>http://purl.obolibrary.org/obo/CL_0000038</td>\n",
       "      <td>erythroid progenitor cell</td>\n",
       "      <td>http://purl.obolibrary.org/obo/CL_0000038</td>\n",
       "      <td>3929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4147605</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>Sample2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>late embryonic stage</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0007220</td>\n",
       "      <td>17 week</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000383</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bone marrow</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002371</td>\n",
       "      <td>hip</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0001464</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4147606</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>Sample2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>late embryonic stage</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0007220</td>\n",
       "      <td>17 week</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000383</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bone marrow</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002371</td>\n",
       "      <td>hip</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0001464</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4147607</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>Sample2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>late embryonic stage</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0007220</td>\n",
       "      <td>17 week</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000383</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bone marrow</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002371</td>\n",
       "      <td>hip</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0001464</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4147608</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>Sample2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>late embryonic stage</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0007220</td>\n",
       "      <td>17 week</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000383</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bone marrow</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002371</td>\n",
       "      <td>hip</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0001464</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4147609</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>Sample2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>late embryonic stage</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0007220</td>\n",
       "      <td>17 week</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000383</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bone marrow</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002371</td>\n",
       "      <td>hip</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0001464</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4147610</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>Sample2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>late embryonic stage</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0007220</td>\n",
       "      <td>17 week</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000383</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bone marrow</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002371</td>\n",
       "      <td>hip</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0001464</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERR4147611</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>http://purl.obolibrary.org/obo/NCBITaxon_9606</td>\n",
       "      <td>Sample2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>late embryonic stage</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0007220</td>\n",
       "      <td>17 week</td>\n",
       "      <td>NaN</td>\n",
       "      <td>female</td>\n",
       "      <td>http://purl.obolibrary.org/obo/PATO_0000383</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>bone marrow</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002371</td>\n",
       "      <td>hip</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0001464</td>\n",
       "      <td>cycling committed progenitor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>progenitor cell</td>\n",
       "      <td>http://purl.obolibrary.org/obo/CL_0011026</td>\n",
       "      <td>5067</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Sample Characteristic[organism]  \\\n",
       "ERR4147602                    Homo sapiens   \n",
       "ERR4147603                    Homo sapiens   \n",
       "ERR4147604                    Homo sapiens   \n",
       "ERR4147605                    Homo sapiens   \n",
       "ERR4147606                    Homo sapiens   \n",
       "ERR4147607                    Homo sapiens   \n",
       "ERR4147608                    Homo sapiens   \n",
       "ERR4147609                    Homo sapiens   \n",
       "ERR4147610                    Homo sapiens   \n",
       "ERR4147611                    Homo sapiens   \n",
       "\n",
       "            Sample Characteristic Ontology Term[organism]  \\\n",
       "ERR4147602  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "ERR4147603  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "ERR4147604  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "ERR4147605  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "ERR4147606  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "ERR4147607  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "ERR4147608  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "ERR4147609  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "ERR4147610  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "ERR4147611  http://purl.obolibrary.org/obo/NCBITaxon_9606   \n",
       "\n",
       "           Sample Characteristic[individual]  \\\n",
       "ERR4147602                           Sample2   \n",
       "ERR4147603                           Sample2   \n",
       "ERR4147604                           Sample2   \n",
       "ERR4147605                           Sample2   \n",
       "ERR4147606                           Sample2   \n",
       "ERR4147607                           Sample2   \n",
       "ERR4147608                           Sample2   \n",
       "ERR4147609                           Sample2   \n",
       "ERR4147610                           Sample2   \n",
       "ERR4147611                           Sample2   \n",
       "\n",
       "            Sample Characteristic Ontology Term[individual]  \\\n",
       "ERR4147602                                              NaN   \n",
       "ERR4147603                                              NaN   \n",
       "ERR4147604                                              NaN   \n",
       "ERR4147605                                              NaN   \n",
       "ERR4147606                                              NaN   \n",
       "ERR4147607                                              NaN   \n",
       "ERR4147608                                              NaN   \n",
       "ERR4147609                                              NaN   \n",
       "ERR4147610                                              NaN   \n",
       "ERR4147611                                              NaN   \n",
       "\n",
       "           Sample Characteristic[developmental stage]  \\\n",
       "ERR4147602                       late embryonic stage   \n",
       "ERR4147603                       late embryonic stage   \n",
       "ERR4147604                       late embryonic stage   \n",
       "ERR4147605                       late embryonic stage   \n",
       "ERR4147606                       late embryonic stage   \n",
       "ERR4147607                       late embryonic stage   \n",
       "ERR4147608                       late embryonic stage   \n",
       "ERR4147609                       late embryonic stage   \n",
       "ERR4147610                       late embryonic stage   \n",
       "ERR4147611                       late embryonic stage   \n",
       "\n",
       "           Sample Characteristic Ontology Term[developmental stage]  \\\n",
       "ERR4147602      http://purl.obolibrary.org/obo/UBERON_0007220         \n",
       "ERR4147603      http://purl.obolibrary.org/obo/UBERON_0007220         \n",
       "ERR4147604      http://purl.obolibrary.org/obo/UBERON_0007220         \n",
       "ERR4147605      http://purl.obolibrary.org/obo/UBERON_0007220         \n",
       "ERR4147606      http://purl.obolibrary.org/obo/UBERON_0007220         \n",
       "ERR4147607      http://purl.obolibrary.org/obo/UBERON_0007220         \n",
       "ERR4147608      http://purl.obolibrary.org/obo/UBERON_0007220         \n",
       "ERR4147609      http://purl.obolibrary.org/obo/UBERON_0007220         \n",
       "ERR4147610      http://purl.obolibrary.org/obo/UBERON_0007220         \n",
       "ERR4147611      http://purl.obolibrary.org/obo/UBERON_0007220         \n",
       "\n",
       "           Sample Characteristic[age]  \\\n",
       "ERR4147602                    17 week   \n",
       "ERR4147603                    17 week   \n",
       "ERR4147604                    17 week   \n",
       "ERR4147605                    17 week   \n",
       "ERR4147606                    17 week   \n",
       "ERR4147607                    17 week   \n",
       "ERR4147608                    17 week   \n",
       "ERR4147609                    17 week   \n",
       "ERR4147610                    17 week   \n",
       "ERR4147611                    17 week   \n",
       "\n",
       "            Sample Characteristic Ontology Term[age]  \\\n",
       "ERR4147602                                       NaN   \n",
       "ERR4147603                                       NaN   \n",
       "ERR4147604                                       NaN   \n",
       "ERR4147605                                       NaN   \n",
       "ERR4147606                                       NaN   \n",
       "ERR4147607                                       NaN   \n",
       "ERR4147608                                       NaN   \n",
       "ERR4147609                                       NaN   \n",
       "ERR4147610                                       NaN   \n",
       "ERR4147611                                       NaN   \n",
       "\n",
       "           Sample Characteristic[sex]  \\\n",
       "ERR4147602                     female   \n",
       "ERR4147603                     female   \n",
       "ERR4147604                     female   \n",
       "ERR4147605                     female   \n",
       "ERR4147606                     female   \n",
       "ERR4147607                     female   \n",
       "ERR4147608                     female   \n",
       "ERR4147609                     female   \n",
       "ERR4147610                     female   \n",
       "ERR4147611                     female   \n",
       "\n",
       "               Sample Characteristic Ontology Term[sex]  ...  \\\n",
       "ERR4147602  http://purl.obolibrary.org/obo/PATO_0000383  ...   \n",
       "ERR4147603  http://purl.obolibrary.org/obo/PATO_0000383  ...   \n",
       "ERR4147604  http://purl.obolibrary.org/obo/PATO_0000383  ...   \n",
       "ERR4147605  http://purl.obolibrary.org/obo/PATO_0000383  ...   \n",
       "ERR4147606  http://purl.obolibrary.org/obo/PATO_0000383  ...   \n",
       "ERR4147607  http://purl.obolibrary.org/obo/PATO_0000383  ...   \n",
       "ERR4147608  http://purl.obolibrary.org/obo/PATO_0000383  ...   \n",
       "ERR4147609  http://purl.obolibrary.org/obo/PATO_0000383  ...   \n",
       "ERR4147610  http://purl.obolibrary.org/obo/PATO_0000383  ...   \n",
       "ERR4147611  http://purl.obolibrary.org/obo/PATO_0000383  ...   \n",
       "\n",
       "           Factor Value Ontology Term[single cell identifier]  \\\n",
       "ERR4147602                                                NaN   \n",
       "ERR4147603                                                NaN   \n",
       "ERR4147604                                                NaN   \n",
       "ERR4147605                                                NaN   \n",
       "ERR4147606                                                NaN   \n",
       "ERR4147607                                                NaN   \n",
       "ERR4147608                                                NaN   \n",
       "ERR4147609                                                NaN   \n",
       "ERR4147610                                                NaN   \n",
       "ERR4147611                                                NaN   \n",
       "\n",
       "           Factor Value[organism part]  \\\n",
       "ERR4147602                 bone marrow   \n",
       "ERR4147603                 bone marrow   \n",
       "ERR4147604                 bone marrow   \n",
       "ERR4147605                 bone marrow   \n",
       "ERR4147606                 bone marrow   \n",
       "ERR4147607                 bone marrow   \n",
       "ERR4147608                 bone marrow   \n",
       "ERR4147609                 bone marrow   \n",
       "ERR4147610                 bone marrow   \n",
       "ERR4147611                 bone marrow   \n",
       "\n",
       "                Factor Value Ontology Term[organism part]  \\\n",
       "ERR4147602  http://purl.obolibrary.org/obo/UBERON_0002371   \n",
       "ERR4147603  http://purl.obolibrary.org/obo/UBERON_0002371   \n",
       "ERR4147604  http://purl.obolibrary.org/obo/UBERON_0002371   \n",
       "ERR4147605  http://purl.obolibrary.org/obo/UBERON_0002371   \n",
       "ERR4147606  http://purl.obolibrary.org/obo/UBERON_0002371   \n",
       "ERR4147607  http://purl.obolibrary.org/obo/UBERON_0002371   \n",
       "ERR4147608  http://purl.obolibrary.org/obo/UBERON_0002371   \n",
       "ERR4147609  http://purl.obolibrary.org/obo/UBERON_0002371   \n",
       "ERR4147610  http://purl.obolibrary.org/obo/UBERON_0002371   \n",
       "ERR4147611  http://purl.obolibrary.org/obo/UBERON_0002371   \n",
       "\n",
       "           Factor Value[sampling site]  \\\n",
       "ERR4147602                         hip   \n",
       "ERR4147603                         hip   \n",
       "ERR4147604                         hip   \n",
       "ERR4147605                         hip   \n",
       "ERR4147606                         hip   \n",
       "ERR4147607                         hip   \n",
       "ERR4147608                         hip   \n",
       "ERR4147609                         hip   \n",
       "ERR4147610                         hip   \n",
       "ERR4147611                         hip   \n",
       "\n",
       "                Factor Value Ontology Term[sampling site]  \\\n",
       "ERR4147602  http://purl.obolibrary.org/obo/UBERON_0001464   \n",
       "ERR4147603  http://purl.obolibrary.org/obo/UBERON_0001464   \n",
       "ERR4147604  http://purl.obolibrary.org/obo/UBERON_0001464   \n",
       "ERR4147605  http://purl.obolibrary.org/obo/UBERON_0001464   \n",
       "ERR4147606  http://purl.obolibrary.org/obo/UBERON_0001464   \n",
       "ERR4147607  http://purl.obolibrary.org/obo/UBERON_0001464   \n",
       "ERR4147608  http://purl.obolibrary.org/obo/UBERON_0001464   \n",
       "ERR4147609  http://purl.obolibrary.org/obo/UBERON_0001464   \n",
       "ERR4147610  http://purl.obolibrary.org/obo/UBERON_0001464   \n",
       "ERR4147611  http://purl.obolibrary.org/obo/UBERON_0001464   \n",
       "\n",
       "           Factor Value[inferred cell type - authors labels]  \\\n",
       "ERR4147602                                               NaN   \n",
       "ERR4147603                                               NaN   \n",
       "ERR4147604                         erythroid progenitor cell   \n",
       "ERR4147605                                               NaN   \n",
       "ERR4147606                                               NaN   \n",
       "ERR4147607                                               NaN   \n",
       "ERR4147608                                               NaN   \n",
       "ERR4147609                                               NaN   \n",
       "ERR4147610                                               NaN   \n",
       "ERR4147611                      cycling committed progenitor   \n",
       "\n",
       "           Factor Value Ontology Term[inferred cell type - authors labels]  \\\n",
       "ERR4147602                                                NaN                \n",
       "ERR4147603                                                NaN                \n",
       "ERR4147604          http://purl.obolibrary.org/obo/CL_0000038                \n",
       "ERR4147605                                                NaN                \n",
       "ERR4147606                                                NaN                \n",
       "ERR4147607                                                NaN                \n",
       "ERR4147608                                                NaN                \n",
       "ERR4147609                                                NaN                \n",
       "ERR4147610                                                NaN                \n",
       "ERR4147611                                                NaN                \n",
       "\n",
       "            Factor Value[inferred cell type - ontology labels]  \\\n",
       "ERR4147602                                                NaN    \n",
       "ERR4147603                                                NaN    \n",
       "ERR4147604                          erythroid progenitor cell    \n",
       "ERR4147605                                                NaN    \n",
       "ERR4147606                                                NaN    \n",
       "ERR4147607                                                NaN    \n",
       "ERR4147608                                                NaN    \n",
       "ERR4147609                                                NaN    \n",
       "ERR4147610                                                NaN    \n",
       "ERR4147611                                    progenitor cell    \n",
       "\n",
       "           Factor Value Ontology Term[inferred cell type - ontology labels]  \\\n",
       "ERR4147602                                                NaN                 \n",
       "ERR4147603                                                NaN                 \n",
       "ERR4147604          http://purl.obolibrary.org/obo/CL_0000038                 \n",
       "ERR4147605                                                NaN                 \n",
       "ERR4147606                                                NaN                 \n",
       "ERR4147607                                                NaN                 \n",
       "ERR4147608                                                NaN                 \n",
       "ERR4147609                                                NaN                 \n",
       "ERR4147610                                                NaN                 \n",
       "ERR4147611          http://purl.obolibrary.org/obo/CL_0011026                 \n",
       "\n",
       "            n_genes  \n",
       "ERR4147602     2191  \n",
       "ERR4147603     1810  \n",
       "ERR4147604     3929  \n",
       "ERR4147605     5115  \n",
       "ERR4147606     2137  \n",
       "ERR4147607     2561  \n",
       "ERR4147608     2890  \n",
       "ERR4147609     2271  \n",
       "ERR4147610     3252  \n",
       "ERR4147611     5067  \n",
       "\n",
       "[10 rows x 33 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata.head(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9c0fb47",
   "metadata": {},
   "source": [
    "Next, we need to load pathway annotation information. \n",
    "\n",
    "In the next cell we use the MIPath `parse_gmt` function to parse a `.gmt` file downloaded from **MSigDB**, and translate it from HUGO gene names to Ensembl gene IDs using a file created using [BioMart](https://www.ensembl.org/info/data/biomart/index.html).\n",
    "\n",
    "For Reactome, MIPath includes a function `get_reactome` that automatically downloads the latest version of the database and parses it. Reactome is already annotated using Ensembl gene IDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d4e16e72",
   "metadata": {},
   "outputs": [],
   "source": [
    "kegg = mipath.parse_gmt(gmt_path = './data/c2.cp.kegg.v7.5.1.symbols.gmt', gene_id_dict_path = './data/name2ensembl_human.txt')\n",
    "\n",
    "reactome = mipath.get_reactome(organism = 'HSA', gene_anot = 'Ensembl')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea17d5a1",
   "metadata": {},
   "source": [
    "We finally have everything we need to run the analysis. There are 2 steps in the MIPath pipeline.\n",
    "\n",
    "The first step is to decompose the data into different modules for each pathway in the dataset. We use the `decompose_pathways` function, inputting both the gene expression data, and the pathway annotation. This is the most computationally expensive part of the analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b6a064f0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Performing pathway decomposition\n",
      "(1/186) --- KEGG_N_GLYCAN_BIOSYNTHESIS --- 46 genes\n",
      "Computed 25 NN in 7.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 23 partitions in 0.3 s\n",
      "\n",
      "(2/186) --- KEGG_OTHER_GLYCAN_DEGRADATION --- 24 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 28 partitions in 0.2 s\n",
      "\n",
      "(3/186) --- KEGG_O_GLYCAN_BIOSYNTHESIS --- 30 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 29 partitions in 0.1 s\n",
      "\n",
      "(4/186) --- KEGG_GLYCOSAMINOGLYCAN_DEGRADATION --- 20 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 30 partitions in 0.1 s\n",
      "\n",
      "(5/186) --- KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_KERATAN_SULFATE --- 14 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 28 partitions in 0.1 s\n",
      "\n",
      "(6/186) --- KEGG_GLYCEROLIPID_METABOLISM --- 57 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 28 partitions in 0.3 s\n",
      "\n",
      "(7/186) --- KEGG_GLYCOSYLPHOSPHATIDYLINOSITOL_GPI_ANCHOR_BIOSYNTHESIS --- 25 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 25 partitions in 0.3 s\n",
      "\n",
      "(8/186) --- KEGG_GLYCEROPHOSPHOLIPID_METABOLISM --- 95 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.1 s\n",
      "Found 24 partitions in 0.2 s\n",
      "\n",
      "(9/186) --- KEGG_ETHER_LIPID_METABOLISM --- 34 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 29 partitions in 0.1 s\n",
      "\n",
      "(10/186) --- KEGG_ARACHIDONIC_ACID_METABOLISM --- 62 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 23 partitions in 0.2 s\n",
      "\n",
      "(11/186) --- KEGG_LINOLEIC_ACID_METABOLISM --- 29 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 34 partitions in 0.2 s\n",
      "\n",
      "(12/186) --- KEGG_ALPHA_LINOLENIC_ACID_METABOLISM --- 19 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 35 partitions in 0.1 s\n",
      "\n",
      "(13/186) --- KEGG_SPHINGOLIPID_METABOLISM --- 48 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 25 partitions in 0.2 s\n",
      "\n",
      "(14/186) --- KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_LACTO_AND_NEOLACTO_SERIES --- 27 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 32 partitions in 0.1 s\n",
      "\n",
      "(15/186) --- KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_GLOBO_SERIES --- 13 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 28 partitions in 0.1 s\n",
      "\n",
      "(16/186) --- KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_GANGLIO_SERIES --- 18 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 30 partitions in 0.2 s\n",
      "\n",
      "(17/186) --- KEGG_RIBOFLAVIN_METABOLISM --- 15 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 32 partitions in 0.1 s\n",
      "\n",
      "(18/186) --- KEGG_NICOTINATE_AND_NICOTINAMIDE_METABOLISM --- 23 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 21 partitions in 0.1 s\n",
      "\n",
      "(19/186) --- KEGG_PANTOTHENATE_AND_COA_BIOSYNTHESIS --- 16 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 30 partitions in 0.2 s\n",
      "\n",
      "(20/186) --- KEGG_AMINOACYL_TRNA_BIOSYNTHESIS --- 53 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 23 partitions in 0.3 s\n",
      "\n",
      "(21/186) --- KEGG_BASAL_TRANSCRIPTION_FACTORS --- 44 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 24 partitions in 0.2 s\n",
      "\n",
      "(22/186) --- KEGG_NON_HOMOLOGOUS_END_JOINING --- 12 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 27 partitions in 0.2 s\n",
      "\n",
      "(23/186) --- KEGG_SNARE_INTERACTIONS_IN_VESICULAR_TRANSPORT --- 37 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.1 s\n",
      "Found 30 partitions in 0.3 s\n",
      "\n",
      "(24/186) --- KEGG_LYSOSOME --- 137 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.2 s\n",
      "Found 10 partitions in 0.2 s\n",
      "\n",
      "(25/186) --- KEGG_CARDIAC_MUSCLE_CONTRACTION --- 81 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 9 partitions in 0.3 s\n",
      "\n",
      "(26/186) --- KEGG_RENIN_ANGIOTENSIN_SYSTEM --- 16 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 27 partitions in 0.2 s\n",
      "\n",
      "(27/186) --- KEGG_CIRCADIAN_RHYTHM_MAMMAL --- 12 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 34 partitions in 0.1 s\n",
      "\n",
      "(28/186) --- KEGG_TASTE_TRANSDUCTION --- 78 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 24 partitions in 0.1 s\n",
      "\n",
      "(29/186) --- KEGG_PROXIMAL_TUBULE_BICARBONATE_RECLAMATION --- 24 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 30 partitions in 0.1 s\n",
      "\n",
      "(30/186) --- KEGG_PATHOGENIC_ESCHERICHIA_COLI_INFECTION --- 67 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 9 partitions in 0.2 s\n",
      "\n",
      "(31/186) --- KEGG_SYSTEMIC_LUPUS_ERYTHEMATOSUS --- 237 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.1 s\n",
      "Found 15 partitions in 0.2 s\n",
      "\n",
      "(32/186) --- KEGG_PRIMARY_IMMUNODEFICIENCY --- 49 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 21 partitions in 0.2 s\n",
      "\n",
      "(33/186) --- KEGG_HYPERTROPHIC_CARDIOMYOPATHY_HCM --- 91 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 16 partitions in 0.3 s\n",
      "\n",
      "(34/186) --- KEGG_ARRHYTHMOGENIC_RIGHT_VENTRICULAR_CARDIOMYOPATHY_ARVC --- 76 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 18 partitions in 0.2 s\n",
      "\n",
      "(35/186) --- KEGG_DILATED_CARDIOMYOPATHY --- 100 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 16 partitions in 0.3 s\n",
      "\n",
      "(36/186) --- KEGG_GLYCOLYSIS_GLUCONEOGENESIS --- 65 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 11 partitions in 0.4 s\n",
      "\n",
      "(37/186) --- KEGG_CITRATE_CYCLE_TCA_CYCLE --- 31 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 22 partitions in 0.3 s\n",
      "\n",
      "(38/186) --- KEGG_PENTOSE_PHOSPHATE_PATHWAY --- 26 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 20 partitions in 0.3 s\n",
      "\n",
      "(39/186) --- KEGG_PENTOSE_AND_GLUCURONATE_INTERCONVERSIONS --- 29 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 30 partitions in 0.2 s\n",
      "\n",
      "(40/186) --- KEGG_FRUCTOSE_AND_MANNOSE_METABOLISM --- 34 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 25 partitions in 0.2 s\n",
      "\n",
      "(41/186) --- KEGG_GALACTOSE_METABOLISM --- 27 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 25 partitions in 0.2 s\n",
      "\n",
      "(42/186) --- KEGG_ASCORBATE_AND_ALDARATE_METABOLISM --- 26 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 36 partitions in 0.2 s\n",
      "\n",
      "(43/186) --- KEGG_FATTY_ACID_METABOLISM --- 41 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 26 partitions in 0.3 s\n",
      "\n",
      "(44/186) --- KEGG_STEROID_BIOSYNTHESIS --- 18 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 27 partitions in 0.1 s\n",
      "\n",
      "(45/186) --- KEGG_PRIMARY_BILE_ACID_BIOSYNTHESIS --- 16 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 29 partitions in 0.1 s\n",
      "\n",
      "(46/186) --- KEGG_STEROID_HORMONE_BIOSYNTHESIS --- 67 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 28 partitions in 0.1 s\n",
      "\n",
      "(47/186) --- KEGG_OXIDATIVE_PHOSPHORYLATION --- 155 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.3 s\n",
      "Found 8 partitions in 0.1 s\n",
      "\n",
      "(48/186) --- KEGG_PURINE_METABOLISM --- 164 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.3 s\n",
      "Found 9 partitions in 0.4 s\n",
      "\n",
      "(49/186) --- KEGG_PYRIMIDINE_METABOLISM --- 99 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.2 s\n",
      "Found 14 partitions in 0.2 s\n",
      "\n",
      "(50/186) --- KEGG_ALANINE_ASPARTATE_AND_GLUTAMATE_METABOLISM --- 32 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 28 partitions in 0.4 s\n",
      "\n",
      "(51/186) --- KEGG_GLYCINE_SERINE_AND_THREONINE_METABOLISM --- 31 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 28 partitions in 0.2 s\n",
      "\n",
      "(52/186) --- KEGG_CYSTEINE_AND_METHIONINE_METABOLISM --- 35 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 21 partitions in 0.3 s\n",
      "\n",
      "(53/186) --- KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_DEGRADATION --- 45 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 22 partitions in 0.2 s\n",
      "\n",
      "(54/186) --- KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_BIOSYNTHESIS --- 21 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 26 partitions in 0.1 s\n",
      "\n",
      "(55/186) --- KEGG_LYSINE_DEGRADATION --- 49 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 24 partitions in 0.2 s\n",
      "\n",
      "(56/186) --- KEGG_ARGININE_AND_PROLINE_METABOLISM --- 55 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 23 partitions in 0.2 s\n",
      "\n",
      "(57/186) --- KEGG_HISTIDINE_METABOLISM --- 29 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 28 partitions in 0.2 s\n",
      "\n",
      "(58/186) --- KEGG_TYROSINE_METABOLISM --- 43 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 25 partitions in 0.3 s\n",
      "\n",
      "(59/186) --- KEGG_PHENYLALANINE_METABOLISM --- 18 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 26 partitions in 0.1 s\n",
      "\n",
      "(60/186) --- KEGG_TRYPTOPHAN_METABOLISM --- 40 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 27 partitions in 0.1 s\n",
      "\n",
      "(61/186) --- KEGG_BETA_ALANINE_METABOLISM --- 21 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 28 partitions in 0.4 s\n",
      "\n",
      "(62/186) --- KEGG_TAURINE_AND_HYPOTAURINE_METABOLISM --- 10 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 54 partitions in 0.1 s\n",
      "\n",
      "(63/186) --- KEGG_SELENOAMINO_ACID_METABOLISM --- 25 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 28 partitions in 0.2 s\n",
      "\n",
      "(64/186) --- KEGG_GLUTATHIONE_METABOLISM --- 52 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 17 partitions in 0.4 s\n",
      "\n",
      "(65/186) --- KEGG_STARCH_AND_SUCROSE_METABOLISM --- 56 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.1 s\n",
      "Found 30 partitions in 0.2 s\n",
      "\n",
      "(66/186) --- KEGG_AMINO_SUGAR_AND_NUCLEOTIDE_SUGAR_METABOLISM --- 45 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 30 partitions in 0.3 s\n",
      "\n",
      "(67/186) --- KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_CHONDROITIN_SULFATE --- 22 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 25 partitions in 0.2 s\n",
      "\n",
      "(68/186) --- KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_HEPARAN_SULFATE --- 26 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 25 partitions in 0.1 s\n",
      "\n",
      "(69/186) --- KEGG_INOSITOL_PHOSPHATE_METABOLISM --- 56 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 22 partitions in 0.2 s\n",
      "\n",
      "(70/186) --- KEGG_PYRUVATE_METABOLISM --- 43 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 19 partitions in 0.3 s\n",
      "\n",
      "(71/186) --- KEGG_GLYOXYLATE_AND_DICARBOXYLATE_METABOLISM --- 15 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 27 partitions in 0.2 s\n",
      "\n",
      "(72/186) --- KEGG_PROPANOATE_METABOLISM --- 34 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 22 partitions in 0.2 s\n",
      "\n",
      "(73/186) --- KEGG_BUTANOATE_METABOLISM --- 34 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 27 partitions in 0.2 s\n",
      "\n",
      "(74/186) --- KEGG_ONE_CARBON_POOL_BY_FOLATE --- 19 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 21 partitions in 0.2 s\n",
      "\n",
      "(75/186) --- KEGG_FOLATE_BIOSYNTHESIS --- 10 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 42 partitions in 0.1 s\n",
      "\n",
      "(76/186) --- KEGG_RETINOL_METABOLISM --- 67 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 27 partitions in 0.1 s\n",
      "\n",
      "(77/186) --- KEGG_PORPHYRIN_AND_CHLOROPHYLL_METABOLISM --- 44 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 28 partitions in 0.2 s\n",
      "\n",
      "(78/186) --- KEGG_TERPENOID_BACKBONE_BIOSYNTHESIS --- 14 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 30 partitions in 0.2 s\n",
      "\n",
      "(79/186) --- KEGG_LIMONENE_AND_PINENE_DEGRADATION --- 9 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 28 partitions in 0.1 s\n",
      "\n",
      "(80/186) --- KEGG_NITROGEN_METABOLISM --- 23 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 27 partitions in 0.2 s\n",
      "\n",
      "(81/186) --- KEGG_SULFUR_METABOLISM --- 12 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 35 partitions in 0.2 s\n",
      "\n",
      "(82/186) --- KEGG_METABOLISM_OF_XENOBIOTICS_BY_CYTOCHROME_P450 --- 72 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 25 partitions in 0.3 s\n",
      "\n",
      "(83/186) --- KEGG_DRUG_METABOLISM_CYTOCHROME_P450 --- 79 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 27 partitions in 0.3 s\n",
      "\n",
      "(84/186) --- KEGG_DRUG_METABOLISM_OTHER_ENZYMES --- 54 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 30 partitions in 0.2 s\n",
      "\n",
      "(85/186) --- KEGG_BIOSYNTHESIS_OF_UNSATURATED_FATTY_ACIDS --- 22 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 27 partitions in 0.2 s\n",
      "\n",
      "(86/186) --- KEGG_ABC_TRANSPORTERS --- 60 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 29 partitions in 0.2 s\n",
      "\n",
      "(87/186) --- KEGG_RIBOSOME --- 104 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 12 partitions in 0.2 s\n",
      "\n",
      "(88/186) --- KEGG_RNA_DEGRADATION --- 79 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.2 s\n",
      "Found 11 partitions in 0.2 s\n",
      "\n",
      "(89/186) --- KEGG_RNA_POLYMERASE --- 30 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.2 s\n",
      "Found 23 partitions in 0.3 s\n",
      "\n",
      "(90/186) --- KEGG_DNA_REPLICATION --- 35 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 24 partitions in 0.2 s\n",
      "\n",
      "(91/186) --- KEGG_SPLICEOSOME --- 160 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.3 s\n",
      "Found 7 partitions in 0.2 s\n",
      "\n",
      "(92/186) --- KEGG_PROTEASOME --- 64 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.2 s\n",
      "Found 8 partitions in 0.4 s\n",
      "\n",
      "(93/186) --- KEGG_PROTEIN_EXPORT --- 23 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 15 partitions in 0.4 s\n",
      "\n",
      "(94/186) --- KEGG_PPAR_SIGNALING_PATHWAY --- 74 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 23 partitions in 0.3 s\n",
      "\n",
      "(95/186) --- KEGG_BASE_EXCISION_REPAIR --- 34 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 20 partitions in 0.3 s\n",
      "\n",
      "(96/186) --- KEGG_NUCLEOTIDE_EXCISION_REPAIR --- 52 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 24 partitions in 0.3 s\n",
      "\n",
      "(97/186) --- KEGG_MISMATCH_REPAIR --- 23 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 24 partitions in 0.2 s\n",
      "\n",
      "(98/186) --- KEGG_HOMOLOGOUS_RECOMBINATION --- 28 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 25 partitions in 0.2 s\n",
      "\n",
      "(99/186) --- KEGG_MAPK_SIGNALING_PATHWAY --- 307 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.2 s\n",
      "Found 10 partitions in 0.1 s\n",
      "\n",
      "(100/186) --- KEGG_ERBB_SIGNALING_PATHWAY --- 88 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 14 partitions in 0.3 s\n",
      "\n",
      "(101/186) --- KEGG_CALCIUM_SIGNALING_PATHWAY --- 181 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.2 s\n",
      "Found 11 partitions in 0.3 s\n",
      "\n",
      "(102/186) --- KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION --- 303 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.1 s\n",
      "Found 17 partitions in 0.2 s\n",
      "\n",
      "(103/186) --- KEGG_CHEMOKINE_SIGNALING_PATHWAY --- 213 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.3 s\n",
      "Found 11 partitions in 0.2 s\n",
      "\n",
      "(104/186) --- KEGG_PHOSPHATIDYLINOSITOL_SIGNALING_SYSTEM --- 80 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 13 partitions in 0.4 s\n",
      "\n",
      "(105/186) --- KEGG_NEUROACTIVE_LIGAND_RECEPTOR_INTERACTION --- 288 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.2 s\n",
      "Found 19 partitions in 0.2 s\n",
      "\n",
      "(106/186) --- KEGG_CELL_CYCLE --- 129 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.2 s\n",
      "Found 8 partitions in 0.3 s\n",
      "\n",
      "(107/186) --- KEGG_OOCYTE_MEIOSIS --- 118 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.2 s\n",
      "Found 9 partitions in 0.2 s\n",
      "\n",
      "(108/186) --- KEGG_P53_SIGNALING_PATHWAY --- 69 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 18 partitions in 0.2 s\n",
      "\n",
      "(109/186) --- KEGG_UBIQUITIN_MEDIATED_PROTEOLYSIS --- 137 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.4 s\n",
      "Found 9 partitions in 0.2 s\n",
      "\n",
      "(110/186) --- KEGG_REGULATION_OF_AUTOPHAGY --- 34 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 26 partitions in 0.3 s\n",
      "\n",
      "(111/186) --- KEGG_ENDOCYTOSIS --- 243 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.2 s\n",
      "Found 12 partitions in 0.2 s\n",
      "\n",
      "(112/186) --- KEGG_PEROXISOME --- 83 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 15 partitions in 0.3 s\n",
      "\n",
      "(113/186) --- KEGG_MTOR_SIGNALING_PATHWAY --- 53 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 18 partitions in 0.3 s\n",
      "\n",
      "(114/186) --- KEGG_APOPTOSIS --- 97 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 15 partitions in 0.3 s\n",
      "\n",
      "(115/186) --- KEGG_VASCULAR_SMOOTH_MUSCLE_CONTRACTION --- 119 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.2 s\n",
      "Found 12 partitions in 0.3 s\n",
      "\n",
      "(116/186) --- KEGG_WNT_SIGNALING_PATHWAY --- 164 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.2 s\n",
      "Found 9 partitions in 0.3 s\n",
      "\n",
      "(117/186) --- KEGG_DORSO_VENTRAL_AXIS_FORMATION --- 31 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 27 partitions in 0.2 s\n",
      "\n",
      "(118/186) --- KEGG_NOTCH_SIGNALING_PATHWAY --- 56 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 25 partitions in 0.2 s\n",
      "\n",
      "(119/186) --- KEGG_HEDGEHOG_SIGNALING_PATHWAY --- 59 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 26 partitions in 0.1 s\n",
      "\n",
      "(120/186) --- KEGG_TGF_BETA_SIGNALING_PATHWAY --- 94 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.1 s\n",
      "Found 16 partitions in 0.3 s\n",
      "\n",
      "(121/186) --- KEGG_AXON_GUIDANCE --- 133 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.1 s\n",
      "Found 10 partitions in 0.3 s\n",
      "\n",
      "(122/186) --- KEGG_VEGF_SIGNALING_PATHWAY --- 79 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.2 s\n",
      "Found 18 partitions in 0.3 s\n",
      "\n",
      "(123/186) --- KEGG_FOCAL_ADHESION --- 217 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.3 s\n",
      "Found 9 partitions in 0.2 s\n",
      "\n",
      "(124/186) --- KEGG_ECM_RECEPTOR_INTERACTION --- 105 genes\n",
      "Computed 25 NN in 0.2 s\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNN scores in 0.1 s\n",
      "Found 24 partitions in 0.2 s\n",
      "\n",
      "(125/186) --- KEGG_CELL_ADHESION_MOLECULES_CAMS --- 250 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.1 s\n",
      "Found 19 partitions in 0.2 s\n",
      "\n",
      "(126/186) --- KEGG_ADHERENS_JUNCTION --- 80 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 9 partitions in 0.3 s\n",
      "\n",
      "(127/186) --- KEGG_TIGHT_JUNCTION --- 149 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.2 s\n",
      "Found 9 partitions in 0.2 s\n",
      "\n",
      "(128/186) --- KEGG_GAP_JUNCTION --- 103 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 15 partitions in 0.2 s\n",
      "\n",
      "(129/186) --- KEGG_COMPLEMENT_AND_COAGULATION_CASCADES --- 91 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 28 partitions in 0.2 s\n",
      "\n",
      "(130/186) --- KEGG_ANTIGEN_PROCESSING_AND_PRESENTATION --- 521 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.1 s\n",
      "Found 16 partitions in 0.2 s\n",
      "\n",
      "(131/186) --- KEGG_TOLL_LIKE_RECEPTOR_SIGNALING_PATHWAY --- 115 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 12 partitions in 0.3 s\n",
      "\n",
      "(132/186) --- KEGG_NOD_LIKE_RECEPTOR_SIGNALING_PATHWAY --- 72 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 14 partitions in 0.3 s\n",
      "\n",
      "(133/186) --- KEGG_RIG_I_LIKE_RECEPTOR_SIGNALING_PATHWAY --- 80 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 23 partitions in 0.3 s\n",
      "\n",
      "(134/186) --- KEGG_CYTOSOLIC_DNA_SENSING_PATHWAY --- 63 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 20 partitions in 0.2 s\n",
      "\n",
      "(135/186) --- KEGG_JAK_STAT_SIGNALING_PATHWAY --- 159 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.2 s\n",
      "Found 14 partitions in 0.2 s\n",
      "\n",
      "(136/186) --- KEGG_HEMATOPOIETIC_CELL_LINEAGE --- 103 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 17 partitions in 0.2 s\n",
      "\n",
      "(137/186) --- KEGG_NATURAL_KILLER_CELL_MEDIATED_CYTOTOXICITY --- 453 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.2 s\n",
      "Found 13 partitions in 0.2 s\n",
      "\n",
      "(138/186) --- KEGG_T_CELL_RECEPTOR_SIGNALING_PATHWAY --- 121 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.2 s\n",
      "Found 8 partitions in 0.3 s\n",
      "\n",
      "(139/186) --- KEGG_B_CELL_RECEPTOR_SIGNALING_PATHWAY --- 82 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 9 partitions in 0.2 s\n",
      "\n",
      "(140/186) --- KEGG_FC_EPSILON_RI_SIGNALING_PATHWAY --- 89 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 15 partitions in 0.4 s\n",
      "\n",
      "(141/186) --- KEGG_FC_GAMMA_R_MEDIATED_PHAGOCYTOSIS --- 100 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.2 s\n",
      "Found 9 partitions in 0.3 s\n",
      "\n",
      "(142/186) --- KEGG_LEUKOCYTE_TRANSENDOTHELIAL_MIGRATION --- 121 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.2 s\n",
      "Found 11 partitions in 0.2 s\n",
      "\n",
      "(143/186) --- KEGG_INTESTINAL_IMMUNE_NETWORK_FOR_IGA_PRODUCTION --- 114 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.1 s\n",
      "Found 17 partitions in 0.1 s\n",
      "\n",
      "(144/186) --- KEGG_LONG_TERM_POTENTIATION --- 73 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 14 partitions in 0.4 s\n",
      "\n",
      "(145/186) --- KEGG_NEUROTROPHIN_SIGNALING_PATHWAY --- 130 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.3 s\n",
      "Found 7 partitions in 0.3 s\n",
      "\n",
      "(146/186) --- KEGG_LONG_TERM_DEPRESSION --- 73 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 17 partitions in 0.3 s\n",
      "\n",
      "(147/186) --- KEGG_OLFACTORY_TRANSDUCTION --- 515 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.1 s\n",
      "Found 17 partitions in 0.2 s\n",
      "\n",
      "(148/186) --- KEGG_REGULATION_OF_ACTIN_CYTOSKELETON --- 219 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.4 s\n",
      "Found 8 partitions in 0.2 s\n",
      "\n",
      "(149/186) --- KEGG_INSULIN_SIGNALING_PATHWAY --- 154 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.3 s\n",
      "Found 8 partitions in 0.3 s\n",
      "\n",
      "(150/186) --- KEGG_GNRH_SIGNALING_PATHWAY --- 105 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 13 partitions in 0.4 s\n",
      "\n",
      "(151/186) --- KEGG_PROGESTERONE_MEDIATED_OOCYTE_MATURATION --- 89 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 13 partitions in 0.3 s\n",
      "\n",
      "(152/186) --- KEGG_MELANOGENESIS --- 106 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 14 partitions in 0.2 s\n",
      "\n",
      "(153/186) --- KEGG_ADIPOCYTOKINE_SIGNALING_PATHWAY --- 84 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.2 s\n",
      "Found 24 partitions in 0.3 s\n",
      "\n",
      "(154/186) --- KEGG_TYPE_II_DIABETES_MELLITUS --- 56 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 28 partitions in 0.1 s\n",
      "\n",
      "(155/186) --- KEGG_TYPE_I_DIABETES_MELLITUS --- 159 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.1 s\n",
      "Found 19 partitions in 0.2 s\n",
      "\n",
      "(156/186) --- KEGG_MATURITY_ONSET_DIABETES_OF_THE_YOUNG --- 25 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 34 partitions in 0.1 s\n",
      "\n",
      "(157/186) --- KEGG_ALDOSTERONE_REGULATED_SODIUM_REABSORPTION --- 41 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 27 partitions in 0.2 s\n",
      "\n",
      "(158/186) --- KEGG_VASOPRESSIN_REGULATED_WATER_REABSORPTION --- 46 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 17 partitions in 0.4 s\n",
      "\n",
      "(159/186) --- KEGG_ALZHEIMERS_DISEASE --- 191 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.3 s\n",
      "Found 7 partitions in 0.2 s\n",
      "\n",
      "(160/186) --- KEGG_PARKINSONS_DISEASE --- 146 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.3 s\n",
      "Found 8 partitions in 0.1 s\n",
      "\n",
      "(161/186) --- KEGG_AMYOTROPHIC_LATERAL_SCLEROSIS_ALS --- 64 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 18 partitions in 0.4 s\n",
      "\n",
      "(162/186) --- KEGG_HUNTINGTONS_DISEASE --- 202 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.3 s\n",
      "Found 6 partitions in 0.2 s\n",
      "\n",
      "(163/186) --- KEGG_PRION_DISEASES --- 40 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 25 partitions in 0.2 s\n",
      "\n",
      "(164/186) --- KEGG_VIBRIO_CHOLERAE_INFECTION --- 65 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 11 partitions in 0.3 s\n",
      "\n",
      "(165/186) --- KEGG_EPITHELIAL_CELL_SIGNALING_IN_HELICOBACTER_PYLORI_INFECTION --- 77 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 16 partitions in 0.2 s\n",
      "\n",
      "(166/186) --- KEGG_LEISHMANIA_INFECTION --- 147 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 16 partitions in 0.2 s\n",
      "\n",
      "(167/186) --- KEGG_PATHWAYS_IN_CANCER --- 337 genes\n",
      "Computed 25 NN in 0.2 s\n",
      "SNN scores in 0.3 s\n",
      "Found 10 partitions in 0.2 s\n",
      "\n",
      "(168/186) --- KEGG_COLORECTAL_CANCER --- 62 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 11 partitions in 0.2 s\n",
      "\n",
      "(169/186) --- KEGG_RENAL_CELL_CARCINOMA --- 71 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 15 partitions in 0.4 s\n",
      "\n",
      "(170/186) --- KEGG_PANCREATIC_CANCER --- 71 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 15 partitions in 0.3 s\n",
      "\n",
      "(171/186) --- KEGG_ENDOMETRIAL_CANCER --- 54 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 23 partitions in 0.3 s\n",
      "\n",
      "(172/186) --- KEGG_GLIOMA --- 68 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 12 partitions in 0.3 s\n",
      "\n",
      "(173/186) --- KEGG_PROSTATE_CANCER --- 92 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 11 partitions in 0.3 s\n",
      "\n",
      "(174/186) --- KEGG_THYROID_CANCER --- 34 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 22 partitions in 0.2 s\n",
      "\n",
      "(175/186) --- KEGG_BASAL_CELL_CARCINOMA --- 57 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.0 s\n",
      "Found 27 partitions in 0.1 s\n",
      "\n",
      "(176/186) --- KEGG_MELANOMA --- 75 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 27 partitions in 0.3 s\n",
      "\n",
      "(177/186) --- KEGG_BLADDER_CANCER --- 43 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 25 partitions in 0.2 s\n",
      "\n",
      "(178/186) --- KEGG_CHRONIC_MYELOID_LEUKEMIA --- 75 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 13 partitions in 0.3 s\n",
      "\n",
      "(179/186) --- KEGG_ACUTE_MYELOID_LEUKEMIA --- 58 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 20 partitions in 0.2 s\n",
      "\n",
      "(180/186) --- KEGG_SMALL_CELL_LUNG_CANCER --- 91 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 16 partitions in 0.3 s\n",
      "\n",
      "(181/186) --- KEGG_NON_SMALL_CELL_LUNG_CANCER --- 61 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 18 partitions in 0.2 s\n",
      "\n",
      "(182/186) --- KEGG_ASTHMA --- 102 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 20 partitions in 0.2 s\n",
      "\n",
      "(183/186) --- KEGG_AUTOIMMUNE_THYROID_DISEASE --- 155 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 19 partitions in 0.2 s\n",
      "\n",
      "(184/186) --- KEGG_ALLOGRAFT_REJECTION --- 146 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 20 partitions in 0.2 s\n",
      "\n",
      "(185/186) --- KEGG_GRAFT_VERSUS_HOST_DISEASE --- 302 genes\n",
      "Computed 25 NN in 0.1 s\n",
      "SNN scores in 0.1 s\n",
      "Found 19 partitions in 0.3 s\n",
      "\n",
      "(186/186) --- KEGG_VIRAL_MYOCARDITIS --- 173 genes\n",
      "Computed 25 NN in 0.1 s\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNN scores in 0.1 s\n",
      "Found 15 partitions in 0.2 s\n",
      "\n"
     ]
    }
   ],
   "source": [
    "decomposed_df = mipath.decompose_pathways(data = data, gene_sets_df = kegg)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6704194c",
   "metadata": {},
   "source": [
    "Now we are ready to see how well these modules represent phenotype information, using the `score_factors` function.\n",
    "\n",
    "We input the dataframe obtained in the previous step, the one containing the phenotype information, and a list of things to score, that must match exactly column names from the phenotype dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7da3c218",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Scoring Factor Value[inferred cell type - authors labels]\n",
      "Scoring Factor Value[sampling site]\n"
     ]
    }
   ],
   "source": [
    "to_score = ['Factor Value[inferred cell type - authors labels]', 'Factor Value[sampling site]']\n",
    "\n",
    "result = mipath.score_factors(decomposed_df, metadata, to_score)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cac18a9",
   "metadata": {},
   "source": [
    "And we can now display the final results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ab90532f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Factor Value[inferred cell type - authors labels]</th>\n",
       "      <th>Factor Value[sampling site]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KEGG_HEMATOPOIETIC_CELL_LINEAGE</th>\n",
       "      <td>0.404919</td>\n",
       "      <td>0.027009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION</th>\n",
       "      <td>0.330273</td>\n",
       "      <td>0.011363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_LYSOSOME</th>\n",
       "      <td>0.313702</td>\n",
       "      <td>0.014283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_CELL_ADHESION_MOLECULES_CAMS</th>\n",
       "      <td>0.303012</td>\n",
       "      <td>0.062967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_ANTIGEN_PROCESSING_AND_PRESENTATION</th>\n",
       "      <td>0.300463</td>\n",
       "      <td>0.074172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_LINOLEIC_ACID_METABOLISM</th>\n",
       "      <td>0.054820</td>\n",
       "      <td>0.002627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_OTHER_GLYCAN_DEGRADATION</th>\n",
       "      <td>0.053092</td>\n",
       "      <td>0.005065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_REGULATION_OF_AUTOPHAGY</th>\n",
       "      <td>0.050851</td>\n",
       "      <td>0.003167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_CIRCADIAN_RHYTHM_MAMMAL</th>\n",
       "      <td>0.044396</td>\n",
       "      <td>0.004115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KEGG_TAURINE_AND_HYPOTAURINE_METABOLISM</th>\n",
       "      <td>0.031721</td>\n",
       "      <td>0.001274</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>186 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Factor Value[inferred cell type - authors labels]  \\\n",
       "KEGG_HEMATOPOIETIC_CELL_LINEAGE                                                       0.404919   \n",
       "KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION                                           0.330273   \n",
       "KEGG_LYSOSOME                                                                         0.313702   \n",
       "KEGG_CELL_ADHESION_MOLECULES_CAMS                                                     0.303012   \n",
       "KEGG_ANTIGEN_PROCESSING_AND_PRESENTATION                                              0.300463   \n",
       "...                                                                                        ...   \n",
       "KEGG_LINOLEIC_ACID_METABOLISM                                                         0.054820   \n",
       "KEGG_OTHER_GLYCAN_DEGRADATION                                                         0.053092   \n",
       "KEGG_REGULATION_OF_AUTOPHAGY                                                          0.050851   \n",
       "KEGG_CIRCADIAN_RHYTHM_MAMMAL                                                          0.044396   \n",
       "KEGG_TAURINE_AND_HYPOTAURINE_METABOLISM                                               0.031721   \n",
       "\n",
       "                                             Factor Value[sampling site]  \n",
       "KEGG_HEMATOPOIETIC_CELL_LINEAGE                                 0.027009  \n",
       "KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION                     0.011363  \n",
       "KEGG_LYSOSOME                                                   0.014283  \n",
       "KEGG_CELL_ADHESION_MOLECULES_CAMS                               0.062967  \n",
       "KEGG_ANTIGEN_PROCESSING_AND_PRESENTATION                        0.074172  \n",
       "...                                                                  ...  \n",
       "KEGG_LINOLEIC_ACID_METABOLISM                                   0.002627  \n",
       "KEGG_OTHER_GLYCAN_DEGRADATION                                   0.005065  \n",
       "KEGG_REGULATION_OF_AUTOPHAGY                                    0.003167  \n",
       "KEGG_CIRCADIAN_RHYTHM_MAMMAL                                    0.004115  \n",
       "KEGG_TAURINE_AND_HYPOTAURINE_METABOLISM                         0.001274  \n",
       "\n",
       "[186 rows x 2 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result.sort_values('Factor Value[inferred cell type - authors labels]', ascending=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
